From 7ce392bbcb43c056a9d7e2b855a90287f714799d Mon Sep 17 00:00:00 2001
From: Gurpreet Ghai <gghai@codeaurora.org>
Date: Fri, 1 Jun 2018 14:28:40 +0530
Subject: [PATCH] BT: Adding APTX Adaptive codec for A2DP source

Adding APTX Adaptive codec and related changes to
A2DP source

CRs-Fixed: 2280722

Change-Id: I0b1fa963180f656b7080de14c952b682da9eae3a
---
 bta/av/bta_av_aact.cc                              |   61 ++
 bta/av/bta_av_api.cc                               |   23 +
 bta/av/bta_av_int.h                                |   14 +
 bta/av/bta_av_main.cc                              |    4 +
 bta/include/bta_av_api.h                           |   13 +
 btif/co/bta_av_co.cc                               |    8 +
 btif/include/btif_av.h                             |    1 +
 include/hardware/bt_av.h                           |    1 +
 stack/Android.bp                                   |    2 +
 stack/a2dp/a2dp_codec_config.cc                    |   11 +
 stack/a2dp/a2dp_vendor.cc                          |   91 ++
 stack/a2dp/a2dp_vendor_aptx_adaptive.cc            | 1157 ++++++++++++++++++++
 stack/a2dp/a2dp_vendor_aptx_adaptive_encoder.cc    |  227 ++++
 stack/include/a2dp_vendor_aptx_adaptive.h          |  173 +++
 .../include/a2dp_vendor_aptx_adaptive_constants.h  |   89 ++
 stack/include/a2dp_vendor_aptx_adaptive_encoder.h  |   80 ++
 16 files changed, 1955 insertions(+)
 create mode 100644 stack/a2dp/a2dp_vendor_aptx_adaptive.cc
 create mode 100644 stack/a2dp/a2dp_vendor_aptx_adaptive_encoder.cc
 create mode 100644 stack/include/a2dp_vendor_aptx_adaptive.h
 create mode 100644 stack/include/a2dp_vendor_aptx_adaptive_constants.h
 create mode 100644 stack/include/a2dp_vendor_aptx_adaptive_encoder.h

diff --git a/bta/av/bta_av_aact.cc b/bta/av/bta_av_aact.cc
index a11dda3..8a0bfbf 100644
--- a/bta/av/bta_av_aact.cc
+++ b/bta/av/bta_av_aact.cc
@@ -126,6 +126,9 @@ static void bta_av_vendor_offload_select_codec(tBTA_AV_SCB* p_scb);
 //static uint8_t bta_av_vendor_offload_convert_sample_rate(uint16_t sample_rate);
 
 void bta_av_vendor_offload_check_stop_start(tBTA_AV_SCB* p_scb);
+void update_sub_band_info(uint8_t **param, int *param_len, uint8_t id, uint16_t data);
+void update_sub_band_info(uint8_t **param, int *param_len, uint8_t id, uint8_t *data, uint8_t len);
+void enc_mode_change_callback(tBTM_VSC_CMPL *param);
 
 /* state machine states */
 enum {
@@ -2634,6 +2637,63 @@ void bta_av_reconfig(tBTA_AV_SCB* p_scb, tBTA_AV_DATA* p_data) {
 
 /*******************************************************************************
  *
+ * Function         bta_av_update_enc_mode
+ *
+ * Description      Sends Vendor Specific Command to SoC
+ *                  with new encoder mode.
+ *
+ * Returns          void
+ *
+ ******************************************************************************/
+void bta_av_update_enc_mode(tBTA_AV_DATA* p_data) {
+  uint8_t param[48];
+  uint8_t *p_param;
+  uint8_t *num_sub_band;
+  int param_len = 0;
+  uint16_t enc_mode = p_data->encoder_mode.enc_mode;
+
+  memset(param, 0, 48);
+  p_param = param;
+
+  *p_param++ = VS_QHCI_ENCODER_MODE_CHANGE;
+  param_len++;
+  num_sub_band = p_param++;
+  param_len++;
+
+  update_sub_band_info(&p_param, &param_len, BTA_AV_ENCODER_MODE_CHANGE_ID, enc_mode);
+  *num_sub_band += 1;
+
+  APPL_TRACE_ERROR("VSC stream: opcode: 0x%02x, num: 0x%02x, ID: 0x%02x, len: 0x%02x, Info1: 0x%02x, Info2: 0x%02x", param[0], param[1], param[2], param[3], param[4], param[5]);
+   BTM_VendorSpecificCommand(HCI_VSQC_CONTROLLER_A2DP_OPCODE, param_len,
+                                 param, enc_mode_change_callback);
+}
+
+void update_sub_band_info(uint8_t **param, int *p_param_len, uint8_t id, uint16_t data)
+{
+  uint8_t *p_param = *param;
+  *p_param++ = BTA_AV_ENCODER_MODE_CHANGE_ID;
+  *p_param_len += 1;
+
+  *p_param++ = 2; /* size of uint16_t */
+  *p_param_len += 1;
+
+  UINT16_TO_STREAM(p_param, data);
+  *p_param_len += 2;
+  *param = p_param;
+}
+
+void update_sub_band_info(uint8_t **param, int *p_param_len, uint8_t id, uint8_t *data, uint8_t len)
+{
+  /* For future use */
+}
+
+void enc_mode_change_callback(tBTM_VSC_CMPL *param)
+{
+  APPL_TRACE_ERROR("Received Encoder mode changed Callback from Controller");
+}
+
+/*******************************************************************************
+ *
  * Function         bta_av_data_path
  *
  * Description      Handle stream data path.
@@ -4061,6 +4121,7 @@ void bta_av_offload_req(tBTA_AV_SCB* p_scb, tBTA_AV_DATA* p_data) {
     else if ((strcmp(codec_name,"AAC")) == 0) codec_type = 2;
     else if ((strcmp(codec_name,"aptX")) == 0) codec_type = 8;
     else if ((strcmp(codec_name,"aptX-HD")) == 0) codec_type = 9;
+    else if ((strcmp(codec_name,"aptX-adaptive")) == 0) codec_type = 10;
     else if ((strcmp(codec_name,"LDAC")) == 0) codec_type = 4;
     else if ((strcmp(codec_name,"aptX-TWS")) == 0) codec_type = 11;
     if ((codec_type == 8) || (codec_type == 9) || (codec_type == 4)) {
diff --git a/bta/av/bta_av_api.cc b/bta/av/bta_av_api.cc
index 2f42e55..4bc896a 100644
--- a/bta/av/bta_av_api.cc
+++ b/bta/av/bta_av_api.cc
@@ -384,6 +384,7 @@ void BTA_AvUpdateMaxAVClient(uint8_t max_clients)
   tBTA_AV_MAX_CLIENT *p_buf;
 
   if ((p_buf = (tBTA_AV_MAX_CLIENT *) osi_malloc(sizeof(tBTA_AV_MAX_CLIENT))) != NULL) {
+    APPL_TRACE_DEBUG("%s: gghai: event: %d", __func__, BTA_AV_UPDATE_MAX_AV_CLIENTS_EVT);
     p_buf->hdr.event = BTA_AV_UPDATE_MAX_AV_CLIENTS_EVT;
     p_buf->max_clients = max_clients;
     bta_sys_sendmsg(p_buf);
@@ -423,6 +424,28 @@ void BTA_AvReconfig(tBTA_AV_HNDL hndl, bool suspend, uint8_t sep_info_idx,
 
 /*******************************************************************************
  *
+ * Function         BTA_AvUpdateEncoderMode
+ *
+ * Description      Update current encoder mode to SoC by sending
+ *                  Vendor Specific Command. It is called based on
+ *                  Encoder feedback of Low Latency and High Quality
+ *                  modes
+ * Returns          void
+ *
+ ******************************************************************************/
+void BTA_AvUpdateEncoderMode(uint16_t enc_mode) {
+  tBTA_AV_ENC_MODE* p_buf =
+      (tBTA_AV_ENC_MODE*)osi_malloc(sizeof(tBTA_AV_ENC_MODE));
+
+  p_buf->hdr.event = BTA_AV_UPDATE_ENCODER_MODE_EVT;
+  p_buf->enc_mode = enc_mode;
+
+  bta_sys_sendmsg(p_buf);
+}
+
+
+/*******************************************************************************
+ *
  * Function         BTA_AvProtectReq
  *
  * Description      Send a content protection request.  This function can only
diff --git a/bta/av/bta_av_int.h b/bta/av/bta_av_int.h
index 80e16d4..29f4201 100644
--- a/bta/av/bta_av_int.h
+++ b/bta/av/bta_av_int.h
@@ -144,6 +144,7 @@ enum {
   BTA_AV_UPDATE_MAX_AV_CLIENTS_EVT,
   BTA_AV_ENABLE_MULTICAST_EVT, /* Event for enable and disable multicast */
   BTA_AV_RC_COLLISSION_DETECTED_EVT,
+  BTA_AV_UPDATE_ENCODER_MODE_EVT,
 #if (TWS_ENABLED == TRUE)
   BTA_AV_SET_EARBUD_ROLE_EVT, /* Set TWS earbud role */
   BTA_AV_SET_TWS_DEVICE_EVT, /* Update TWS state */
@@ -171,6 +172,9 @@ enum {
 #define BTA_AV_LAST_EVT BTA_AV_RC_COLLISSION_DETECTED_EVT
 #endif
 
+/* Info ID from updating aptX Adaptive Encoder mode */
+#define BTA_AV_ENCODER_MODE_CHANGE_ID 5
+
 /* maximum number of SEPS in stream discovery results */
 #define BTA_AV_NUM_SEPS 32
 
@@ -382,6 +386,13 @@ typedef struct {
   uint8_t sep_info_idx;
 } tBTA_AV_API_RCFG;
 
+/* data type for BTA_AV_UPDATE_ENCODER_MODE_EVT */
+typedef struct {
+  BT_HDR hdr;
+  uint16_t enc_mode;
+} tBTA_AV_ENC_MODE;
+
+
 /* data type for BTA_AV_CI_SETCONFIG_OK_EVT and BTA_AV_CI_SETCONFIG_FAIL_EVT */
 typedef struct {
   BT_HDR hdr;
@@ -505,6 +516,7 @@ typedef union {
   tBTA_AV_API_STATUS_RSP api_status_rsp;
   tBTA_AV_ENABLE_MULTICAST multicast_state;
   tBTA_AV_MAX_CLIENT max_av_clients;
+  tBTA_AV_ENC_MODE encoder_mode;
 #if (TWS_ENABLED == TRUE)
   tBTA_AV_TWS_SET_EARBUD_ROLE tws_set_earbud_role;
   tBTA_AV_SET_TWS_DEVICE tws_set_device;
@@ -726,6 +738,7 @@ extern tBT_VENDOR_A2DP_OFFLOAD offload_start;
 #define VS_QHCI_A2DP_OFFLOAD_START            0x0A
 #define VS_QHCI_GET_SCRAMBLING_FREQS          0x11
 #define VS_QHCI_SCRAMBLE_A2DP_MEDIA           0x12
+#define VS_QHCI_ENCODER_MODE_CHANGE           0X13
 #define A2DP_TRANSPORT_TYPE_SLIMBUS     0
 #define QHCI_INVALID_VSC 0x01
 /* SPLITA2DP */
@@ -797,6 +810,7 @@ extern void bta_av_rc_disc(uint8_t disc);
 extern void bta_av_conn_chg(tBTA_AV_DATA* p_data);
 extern void bta_av_dereg_comp(tBTA_AV_DATA* p_data);
 extern void bta_av_rc_collission_detected(tBTA_AV_DATA *p_data);
+extern void bta_av_update_enc_mode(tBTA_AV_DATA* p_data);
 
 /* sm action functions */
 extern void bta_av_disable(tBTA_AV_CB* p_cb, tBTA_AV_DATA* p_data);
diff --git a/bta/av/bta_av_main.cc b/bta/av/bta_av_main.cc
index bd0d53f..f6e056c 100644
--- a/bta/av/bta_av_main.cc
+++ b/bta/av/bta_av_main.cc
@@ -200,6 +200,7 @@ extern int btif_max_av_clients;
 static void bta_av_api_set_tws_earbud_role(tBTA_AV_DATA * p_data);
 static void bta_av_api_set_is_tws_device(tBTA_AV_DATA * p_data);
 #endif
+
 /* action functions */
 const tBTA_AV_NSM_ACT bta_av_nsm_act[] = {
     bta_av_api_enable,       /* BTA_AV_API_ENABLE_EVT */
@@ -223,6 +224,7 @@ const tBTA_AV_NSM_ACT bta_av_nsm_act[] = {
     bta_av_api_update_max_av_clients,
     bta_av_api_enable_multicast,    /* BTA_AV_ENABLE_MULTICAST_EVT */
     bta_av_rc_collission_detected, /* BTA_AV_RC_COLLISSION_DETECTED_EVT */
+    bta_av_update_enc_mode, /* BTA_AV_UPDATE_ENCODER_MODE_EVT */
 #if (TWS_ENABLED == TRUE)
     bta_av_api_set_tws_earbud_role, /* BTA_AV_SET_EARBUD_ROLE_EVT */
     bta_av_api_set_is_tws_device, /* BTA_AV_SET_TWS_DEVICE_EVT */
@@ -1650,6 +1652,8 @@ const char* bta_av_evt_code(uint16_t evt_code) {
       return "API_STOP";
     case BTA_AV_ENABLE_MULTICAST_EVT:
       return "MULTICAST_ENABLE";
+    case BTA_AV_UPDATE_ENCODER_MODE_EVT:
+      return "UPDATE_ENCODER_MODE";
     default:
       return "unknown";
   }
diff --git a/bta/include/bta_av_api.h b/bta/include/bta_av_api.h
index 76d1795..efad3ab 100644
--- a/bta/include/bta_av_api.h
+++ b/bta/include/bta_av_api.h
@@ -720,6 +720,19 @@ void BTA_AvReconfig(tBTA_AV_HNDL hndl, bool suspend, uint8_t sep_info_idx,
 
 /*******************************************************************************
  *
+ * Function         BTA_AvUpdateEncoderMode
+ *
+ * Description      Update current encoder mode to SoC by sending
+ *                  Vendor Specific Command. It is called based on
+ *                  Encoder feedback of Low Latency and High Quality
+ *                  modes
+ * Returns          void
+ *
+ ******************************************************************************/
+void BTA_AvUpdateEncoderMode(uint16_t enc_mode);
+
+/*******************************************************************************
+ *
  * Function         BTA_AvProtectReq
  *
  * Description      Send a content protection request.  This function can only
diff --git a/btif/co/bta_av_co.cc b/btif/co/bta_av_co.cc
index 1c9a390..65a5ff2 100644
--- a/btif/co/bta_av_co.cc
+++ b/btif/co/bta_av_co.cc
@@ -1241,6 +1241,7 @@ void bta_av_co_get_peer_params(tA2DP_ENCODER_INIT_PEER_PARAMS* p_peer_params) {
     min_mtu = p_peer->mtu;
     if (min_mtu > BTA_AV_MAX_A2DP_MTU)
         min_mtu = BTA_AV_MAX_A2DP_MTU;
+    if(min_mtu == 0) min_mtu = 672;
     APPL_TRACE_DEBUG("%s updating peer MTU to %d for index %d",
                                     __func__, min_mtu, index);
   }
@@ -1296,6 +1297,13 @@ bool bta_av_co_set_codec_user_config(
     goto done;
   }
 
+  if(codec_user_config.codec_type == BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE
+            && codec_user_config.codec_specific_4 > 0) {
+    APPL_TRACE_DEBUG("%s: Updating Encoder Mode to: %x", __func__, codec_user_config.codec_specific_4);
+    BTA_AvUpdateEncoderMode(codec_user_config.codec_specific_4);
+    return success;
+  }
+
   // Find the peer SEP codec to use
 #if (TWS_ENABLED == TRUE)
   if (codec_user_config.codec_type < BTAV_VENDOR_A2DP_CODEC_INDEX_MAX) {
diff --git a/btif/include/btif_av.h b/btif/include/btif_av.h
index 83719bd..650924f 100644
--- a/btif/include/btif_av.h
+++ b/btif/include/btif_av.h
@@ -62,6 +62,7 @@ typedef enum {
   BTIF_AV_REINIT_AUDIO_IF,
   BTIF_AV_SETUP_CODEC_REQ_EVT,
   BTIF_AV_TRIGGER_HANDOFF_REQ_EVT,
+  BTIF_AV_ENCODER_MODE_CHANGED_EVT,
 } btif_av_sm_event_t;
 
 /*******************************************************************************
diff --git a/include/hardware/bt_av.h b/include/hardware/bt_av.h
index 36722d4..dafd446 100644
--- a/include/hardware/bt_av.h
+++ b/include/hardware/bt_av.h
@@ -53,6 +53,7 @@ typedef enum {
   BTAV_A2DP_CODEC_INDEX_SOURCE_AAC,
   BTAV_A2DP_CODEC_INDEX_SOURCE_APTX,
   BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_HD,
+  BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE,
   BTAV_A2DP_CODEC_INDEX_SOURCE_LDAC,
 
   BTAV_A2DP_CODEC_INDEX_SOURCE_MAX,
diff --git a/stack/Android.bp b/stack/Android.bp
index c609f93..079023d 100644
--- a/stack/Android.bp
+++ b/stack/Android.bp
@@ -54,8 +54,10 @@ cc_library_static {
         "a2dp/a2dp_vendor.cc",
         "a2dp/a2dp_vendor_aptx.cc",
         "a2dp/a2dp_vendor_aptx_hd.cc",
+        "a2dp/a2dp_vendor_aptx_adaptive.cc",
         "a2dp/a2dp_vendor_aptx_encoder.cc",
         "a2dp/a2dp_vendor_aptx_hd_encoder.cc",
+        "a2dp/a2dp_vendor_aptx_adaptive_encoder.cc",
         "a2dp/a2dp_vendor_ldac.cc",
         "a2dp/a2dp_vendor_ldac_abr.cc",
         "a2dp/a2dp_vendor_ldac_encoder.cc",
diff --git a/stack/a2dp/a2dp_codec_config.cc b/stack/a2dp/a2dp_codec_config.cc
index 97f571b..7b51994 100644
--- a/stack/a2dp/a2dp_codec_config.cc
+++ b/stack/a2dp/a2dp_codec_config.cc
@@ -65,6 +65,7 @@
 #include "a2dp_vendor.h"
 #include "a2dp_vendor_aptx.h"
 #include "a2dp_vendor_aptx_hd.h"
+#include "a2dp_vendor_aptx_adaptive.h"
 #include "a2dp_vendor_ldac.h"
 #include "osi/include/log.h"
 #include "a2dp_vendor_aptx_tws.h"
@@ -83,6 +84,7 @@ bool sbc_offload = false;
 bool aac_offload = false;
 bool aptx_offload = false;
 bool aptxhd_offload = false;
+bool aptx_adaptive_offload = false;
 bool ldac_offload = false;
 bool aptxtws_offload = false;
 static void init_btav_a2dp_codec_config(
@@ -166,6 +168,9 @@ A2dpCodecConfig* A2dpCodecConfig::createCodec(
     case BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_HD:
       codec_config = new A2dpCodecConfigAptxHd(codec_priority);
       break;
+    case BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE:
+      codec_config = new A2dpCodecConfigAptxAdaptive(codec_priority);
+      break;
     case BTAV_A2DP_CODEC_INDEX_SOURCE_LDAC:
       codec_config = new A2dpCodecConfigLdac(codec_priority);
       break;
@@ -1466,6 +1471,9 @@ void A2DP_SetOffloadStatus(bool offload_status, char *offload_cap, bool scrambli
       } else if (strcmp(tok,"aptxhd") == 0) {
         LOG_INFO(LOG_TAG,"%s: APTXHD offload supported",__func__);
         aptxhd_offload = TRUE;
+      } else if (strcmp(tok,"aptxadaptive") == 0) {
+        LOG_INFO(LOG_TAG,"%s: APTXHD offload supported",__func__);
+        aptx_adaptive_offload = TRUE;
       } else if (strcmp(tok,"ldac") == 0) {
         LOG_INFO(LOG_TAG,"%s: ldac offload supported",__func__);
         ldac_offload = TRUE;
@@ -1508,6 +1516,9 @@ bool A2DP_IsCodecEnabledInOffload(btav_a2dp_codec_index_t codec_index) {
     case BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_HD:
       codec_status = aptxhd_offload;
       break;
+    case BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE:
+      codec_status = aptx_adaptive_offload;
+      break;
     case BTAV_A2DP_CODEC_INDEX_SOURCE_LDAC:
       if (!ldac_offload)
           LOG_INFO(LOG_TAG,"LDAC not enabled in offload currently");
diff --git a/stack/a2dp/a2dp_vendor.cc b/stack/a2dp/a2dp_vendor.cc
index 36fb761..edde67f 100644
--- a/stack/a2dp/a2dp_vendor.cc
+++ b/stack/a2dp/a2dp_vendor.cc
@@ -61,6 +61,9 @@
 #include "osi/include/log.h"
 #include "osi/include/osi.h"
 #include "a2dp_vendor_aptx_tws.h"
+#include "a2dp_vendor_aptx_adaptive.h"
+#include "a2dp_vendor_aptx_adaptive_constants.h"
+
 #include "bt_vendor_av.h"
 bool A2DP_IsVendorSourceCodecValid(const uint8_t* p_codec_info) {
   uint32_t vendor_id = A2DP_VendorCodecGetVendorId(p_codec_info);
@@ -78,6 +81,12 @@ bool A2DP_IsVendorSourceCodecValid(const uint8_t* p_codec_info) {
     return A2DP_IsVendorSourceCodecValidAptxHd(p_codec_info);
   }
 
+  // Check for aptX-Adaptive
+  if (vendor_id == A2DP_APTX_ADAPTIVE_VENDOR_ID &&
+      codec_id == A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+    return A2DP_IsVendorSourceCodecValidAptxAdaptive(p_codec_info);
+  }
+
   // Check for LDAC
   if (vendor_id == A2DP_LDAC_VENDOR_ID && codec_id == A2DP_LDAC_CODEC_ID) {
     return A2DP_IsVendorSourceCodecValidLdac(p_codec_info);
@@ -132,6 +141,12 @@ bool A2DP_IsVendorPeerSinkCodecValid(const uint8_t* p_codec_info) {
     return A2DP_IsVendorPeerSinkCodecValidAptxHd(p_codec_info);
   }
 
+  // Check for aptX-Adaptive
+  if (vendor_id == A2DP_APTX_ADAPTIVE_VENDOR_ID &&
+      codec_id == A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+    return A2DP_IsVendorPeerSinkCodecValidAptxAdaptive(p_codec_info);
+  }
+
   // Check for LDAC
   if (vendor_id == A2DP_LDAC_VENDOR_ID && codec_id == A2DP_LDAC_CODEC_ID) {
     return A2DP_IsVendorPeerSinkCodecValidLdac(p_codec_info);
@@ -218,6 +233,13 @@ bool A2DP_VendorUsesRtpHeader(bool content_protection_enabled,
                                           p_codec_info);
   }
 
+  // Check for aptX-Adaptive
+  if (vendor_id == A2DP_APTX_ADAPTIVE_VENDOR_ID &&
+      codec_id == A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+    return A2DP_VendorUsesRtpHeaderAptxHd(content_protection_enabled,
+                                          p_codec_info);
+  }
+
   // Check for LDAC
   if (vendor_id == A2DP_LDAC_VENDOR_ID && codec_id == A2DP_LDAC_CODEC_ID) {
     return A2DP_VendorUsesRtpHeaderLdac(content_protection_enabled,
@@ -253,6 +275,13 @@ const char* A2DP_VendorCodecName(UNUSED_ATTR const uint8_t* p_codec_info) {
     return A2DP_VendorCodecNameAptxHd(p_codec_info);
   }
 
+  // Check for aptX-Adaptive
+  if (vendor_id == A2DP_APTX_ADAPTIVE_VENDOR_ID &&
+      codec_id == A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+    return A2DP_VendorCodecNameAptxAdaptive(p_codec_info);
+  }
+
+
   // Check for LDAC
   if (vendor_id == A2DP_LDAC_VENDOR_ID && codec_id == A2DP_LDAC_CODEC_ID) {
     return A2DP_VendorCodecNameLdac(p_codec_info);
@@ -299,6 +328,12 @@ bool A2DP_VendorCodecTypeEquals(const uint8_t* p_codec_info_a,
     return A2DP_VendorCodecTypeEqualsAptxHd(p_codec_info_a, p_codec_info_b);
   }
 
+  // Check for aptX-Adaptive
+  if (vendor_id_a == A2DP_APTX_ADAPTIVE_VENDOR_ID &&
+      codec_id_a == A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+    return A2DP_VendorCodecTypeEqualsAptxAdaptive(p_codec_info_a, p_codec_info_b);
+  }
+
   // Check for LDAC
   if (vendor_id_a == A2DP_LDAC_VENDOR_ID && codec_id_a == A2DP_LDAC_CODEC_ID) {
     return A2DP_VendorCodecTypeEqualsLdac(p_codec_info_a, p_codec_info_b);
@@ -346,6 +381,12 @@ bool A2DP_VendorCodecEquals(const uint8_t* p_codec_info_a,
     return A2DP_VendorCodecEqualsAptxHd(p_codec_info_a, p_codec_info_b);
   }
 
+  // Check for aptX-Adaptive
+  if (vendor_id_a == A2DP_APTX_ADAPTIVE_VENDOR_ID &&
+      codec_id_a == A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+    return A2DP_VendorCodecEqualsAptxAdaptive(p_codec_info_a, p_codec_info_b);
+  }
+
   // Check for LDAC
   if (vendor_id_a == A2DP_LDAC_VENDOR_ID && codec_id_a == A2DP_LDAC_CODEC_ID) {
     return A2DP_VendorCodecEqualsLdac(p_codec_info_a, p_codec_info_b);
@@ -393,6 +434,12 @@ int A2DP_VendorGetTrackSampleRate(const uint8_t* p_codec_info) {
     return A2DP_VendorGetTrackSampleRateAptxHd(p_codec_info);
   }
 
+  // Check for aptX-Adaptive
+  if (vendor_id == A2DP_APTX_ADAPTIVE_VENDOR_ID &&
+      codec_id == A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+    return A2DP_VendorGetTrackSampleRateAptxAdaptive(p_codec_info);
+  }
+
   // Check for LDAC
   if (vendor_id == A2DP_LDAC_VENDOR_ID && codec_id == A2DP_LDAC_CODEC_ID) {
     return A2DP_VendorGetTrackSampleRateLdac(p_codec_info);
@@ -426,6 +473,12 @@ int A2DP_VendorGetTrackChannelCount(const uint8_t* p_codec_info) {
     return A2DP_VendorGetTrackChannelCountAptxHd(p_codec_info);
   }
 
+  // Check for aptX-Adaptive
+  if (vendor_id == A2DP_APTX_ADAPTIVE_VENDOR_ID &&
+      codec_id == A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+    return A2DP_VendorGetTrackChannelCountAptxAdaptive(p_codec_info);
+  }
+
   // Check for LDAC
   if (vendor_id == A2DP_LDAC_VENDOR_ID && codec_id == A2DP_LDAC_CODEC_ID) {
     return A2DP_VendorGetTrackChannelCountLdac(p_codec_info);
@@ -485,6 +538,13 @@ bool A2DP_VendorGetPacketTimestamp(const uint8_t* p_codec_info,
                                                p_timestamp);
   }
 
+  // Check for aptX-Adaptive
+  if (vendor_id == A2DP_APTX_ADAPTIVE_VENDOR_ID &&
+      codec_id == A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+    return A2DP_VendorGetPacketTimestampAptxAdaptive(p_codec_info, p_data,
+                                               p_timestamp);
+  }
+
   // Check for LDAC
   if (vendor_id == A2DP_LDAC_VENDOR_ID && codec_id == A2DP_LDAC_CODEC_ID) {
     return A2DP_VendorGetPacketTimestampLdac(p_codec_info, p_data, p_timestamp);
@@ -514,6 +574,13 @@ bool A2DP_VendorBuildCodecHeader(const uint8_t* p_codec_info, BT_HDR* p_buf,
                                              frames_per_packet);
   }
 
+  // Check for aptX-Adaptive
+  if (vendor_id == A2DP_APTX_ADAPTIVE_VENDOR_ID &&
+      codec_id == A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+    return A2DP_VendorBuildCodecHeaderAptxAdaptive(p_codec_info, p_buf,
+                                             frames_per_packet);
+  }
+
   // Check for LDAC
   if (vendor_id == A2DP_LDAC_VENDOR_ID && codec_id == A2DP_LDAC_CODEC_ID) {
     return A2DP_VendorBuildCodecHeaderLdac(p_codec_info, p_buf,
@@ -542,6 +609,12 @@ const tA2DP_ENCODER_INTERFACE* A2DP_VendorGetEncoderInterface(
     return A2DP_VendorGetEncoderInterfaceAptxHd(p_codec_info);
   }
 
+  // Check for aptX-Adaptive
+  if (vendor_id == A2DP_APTX_ADAPTIVE_VENDOR_ID &&
+      codec_id == A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+    return A2DP_VendorGetEncoderInterfaceAptxAdaptive(p_codec_info);
+  }
+
   // Check for LDAC
   if (vendor_id == A2DP_LDAC_VENDOR_ID && codec_id == A2DP_LDAC_CODEC_ID) {
     return A2DP_VendorGetEncoderInterfaceLdac(p_codec_info);
@@ -575,6 +648,12 @@ bool A2DP_VendorAdjustCodec(uint8_t* p_codec_info) {
     return A2DP_VendorAdjustCodecAptxHd(p_codec_info);
   }
 
+  // Check for aptX-Adaptive
+  if (vendor_id == A2DP_APTX_ADAPTIVE_VENDOR_ID &&
+      codec_id == A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+    return A2DP_VendorAdjustCodecAptxAdaptive(p_codec_info);
+  }
+
   // Check for LDAC
   if (vendor_id == A2DP_LDAC_VENDOR_ID && codec_id == A2DP_LDAC_CODEC_ID) {
     return A2DP_VendorAdjustCodecLdac(p_codec_info);
@@ -597,6 +676,8 @@ btav_a2dp_codec_index_t A2DP_VendorSourceCodecIndex(
   uint32_t vendor_id = A2DP_VendorCodecGetVendorId(p_codec_info);
   uint16_t codec_id = A2DP_VendorCodecGetCodecId(p_codec_info);
 
+   LOG_INFO(LOG_TAG,"A2DP_VendorSourceCodecIndex: vendor_id = %d, codec_id = %d", vendor_id, codec_id);
+
   // Check for aptX
   if (vendor_id == A2DP_APTX_VENDOR_ID &&
       codec_id == A2DP_APTX_CODEC_ID_BLUETOOTH) {
@@ -609,6 +690,12 @@ btav_a2dp_codec_index_t A2DP_VendorSourceCodecIndex(
     return A2DP_VendorSourceCodecIndexAptxHd(p_codec_info);
   }
 
+  // Check for aptX-Adaptive
+  if (vendor_id == A2DP_APTX_ADAPTIVE_VENDOR_ID &&
+      codec_id == A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+    return A2DP_VendorSourceCodecIndexAptxAdaptive(p_codec_info);
+  }
+
   // Check for LDAC
   if (vendor_id == A2DP_LDAC_VENDOR_ID && codec_id == A2DP_LDAC_CODEC_ID) {
     return A2DP_VendorSourceCodecIndexLdac(p_codec_info);
@@ -645,6 +732,8 @@ const char* A2DP_VendorCodecIndexStr(btav_a2dp_codec_index_t codec_index) {
       return A2DP_VendorCodecIndexStrAptx();
     case BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_HD:
       return A2DP_VendorCodecIndexStrAptxHd();
+    case BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE:
+      return A2DP_VendorCodecIndexStrAptxAdaptive();
     case BTAV_A2DP_CODEC_INDEX_SOURCE_LDAC:
       return A2DP_VendorCodecIndexStrLdac();
 #if (TWS_ENABLED == TRUE)
@@ -678,6 +767,8 @@ bool A2DP_VendorInitCodecConfig(btav_a2dp_codec_index_t codec_index,
       return A2DP_VendorInitCodecConfigAptx(p_cfg);
     case BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_HD:
       return A2DP_VendorInitCodecConfigAptxHd(p_cfg);
+    case BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE:
+      return A2DP_VendorInitCodecConfigAptxAdaptive(p_cfg);
     case BTAV_A2DP_CODEC_INDEX_SOURCE_LDAC:
       return A2DP_VendorInitCodecConfigLdac(p_cfg);
 #if (TWS_ENABLED == TRUE)
diff --git a/stack/a2dp/a2dp_vendor_aptx_adaptive.cc b/stack/a2dp/a2dp_vendor_aptx_adaptive.cc
new file mode 100644
index 0000000..a469904
--- /dev/null
+++ b/stack/a2dp/a2dp_vendor_aptx_adaptive.cc
@@ -0,0 +1,1157 @@
+/******************************************************************************
+    Copyright (c) 2018, The Linux Foundation. All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted (subject to the limitations in the
+    disclaimer below) provided that the following conditions are met:
+
+       * Redistributions of source code must retain the above copyright
+         notice, this list of conditions and the following disclaimer.
+
+       * Redistributions in binary form must reproduce the above
+         copyright notice, this list of conditions and the following
+         disclaimer in the documentation and/or other materials provided
+         with the distribution.
+
+       * Neither the name of The Linux Foundation nor the names of its
+         contributors may be used to endorse or promote products derived
+         from this software without specific prior written permission.
+
+    NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE
+    GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT
+    HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
+    WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+    MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+    IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+    ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+    GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+    IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+    OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+    IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+
+/******************************************************************************
+ *
+ *  Utility functions to help build and parse the aptX-adaptive Codec Information
+ *  Element and Media Payload.
+ *
+ ******************************************************************************/
+
+#define LOG_TAG "a2dp_vendor_aptx_adaptive"
+
+#include "bt_target.h"
+
+#include "a2dp_vendor_aptx_adaptive.h"
+
+#include <string.h>
+
+#include <base/logging.h>
+#include "a2dp_vendor.h"
+#include "a2dp_vendor_aptx_adaptive_encoder.h"
+#include "bt_utils.h"
+#include "osi/include/log.h"
+#include "osi/include/osi.h"
+#include "hardware/bt_av.h"
+
+
+// data type for the aptX-adaptive Codec Information Element */
+
+// This is due to be updated
+typedef struct {
+  uint32_t vendorId;
+  uint16_t codecId;    /* Codec ID for aptX-adaptive */
+  uint8_t sampleRate;  /* Sampling Frequency */
+  uint8_t sourceType;  /* AVRCP CUSTOM, Q2Q OR NQ2Q, RESERVED */
+  uint8_t channelMode;
+  uint8_t ttp_ll_0;
+  uint8_t ttp_ll_1;
+  uint8_t ttp_hq_0;
+  uint8_t ttp_hq_1;
+  uint8_t ttp_tws_0;
+  uint8_t ttp_tws_1;
+  uint8_t eoc0;
+  uint8_t eoc1;
+  uint8_t eoc2;
+
+  btav_a2dp_codec_bits_per_sample_t bits_per_sample;
+} tA2DP_APTX_ADAPTIVE_CIE;
+
+/* aptX-adaptive Source codec capabilities */
+static const tA2DP_APTX_ADAPTIVE_CIE a2dp_aptx_adaptive_src_caps = {
+    A2DP_APTX_ADAPTIVE_VENDOR_ID,          /* vendorId */
+    A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH, /* codecId */
+    //A2DP_APTX_ADAPTIVE_SAMPLERATE_48000,   /* sampleRate */
+    A2DP_APTX_ADAPTIVE_SAMPLERATE_48000,   /* sampleRate */
+    A2DP_APTX_ADAPTIVE_SOURCE_TYPE_2,
+    A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO,
+    A2DP_APTX_ADAPTIVE_TTP_LL_0,
+    A2DP_APTX_ADAPTIVE_TTP_LL_1,
+    A2DP_APTX_ADAPTIVE_TTP_HQ_0,
+    A2DP_APTX_ADAPTIVE_TTP_HQ_1,
+    A2DP_APTX_ADAPTIVE_TTP_TWS_0,
+    A2DP_APTX_ADAPTIVE_TTP_TWS_1,
+    A2DP_APTX_ADAPTIVE_EOC0,
+    A2DP_APTX_ADAPTIVE_EOC1,
+    A2DP_APTX_ADAPTIVE_EOC2,
+
+  //  BTAV_A2DP_CODEC_BITS_PER_SAMPLE_24 /* bits_per_sample */
+    BTAV_A2DP_CODEC_BITS_PER_SAMPLE_16 /* bits_per_sample */
+};
+
+/* Default aptX-adaptive codec configuration */
+static const tA2DP_APTX_ADAPTIVE_CIE a2dp_aptx_adaptive_offload_caps = {
+    A2DP_APTX_ADAPTIVE_VENDOR_ID,          /* vendorId */
+    A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH, /* codecId */
+    A2DP_APTX_ADAPTIVE_SAMPLERATE_48000,   /* sampleRate */
+    A2DP_APTX_ADAPTIVE_SOURCE_TYPE_1,
+    A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO | A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO | A2DP_APTX_ADAPTIVE_CHANNELS_TWS_MONO,      /* channelMode */
+    A2DP_APTX_ADAPTIVE_TTP_LL_0,
+    A2DP_APTX_ADAPTIVE_TTP_LL_1,
+    A2DP_APTX_ADAPTIVE_TTP_HQ_0,
+    A2DP_APTX_ADAPTIVE_TTP_HQ_1,
+    A2DP_APTX_ADAPTIVE_TTP_TWS_0,
+    A2DP_APTX_ADAPTIVE_TTP_TWS_1,
+    A2DP_APTX_ADAPTIVE_EOC0,
+    A2DP_APTX_ADAPTIVE_EOC1,
+    A2DP_APTX_ADAPTIVE_EOC2,
+
+    //BTAV_A2DP_CODEC_BITS_PER_SAMPLE_24 /* bits_per_sample */
+    BTAV_A2DP_CODEC_BITS_PER_SAMPLE_16 /* bits_per_sample */
+};
+/* Default aptX-adaptive codec configuration */
+static const tA2DP_APTX_ADAPTIVE_CIE a2dp_aptx_adaptive_default_src_config = {
+//static const tA2DP_APTX_ADAPTIVE_CIE a2dp_aptx_adaptive_default_config = {
+    A2DP_APTX_ADAPTIVE_VENDOR_ID,          /* vendorId */
+    A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH, /* codecId */
+    A2DP_APTX_ADAPTIVE_SAMPLERATE_48000,   /* sampleRate */
+    A2DP_APTX_ADAPTIVE_SOURCE_TYPE_2,
+    A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO,      /* channelMode */
+    A2DP_APTX_ADAPTIVE_TTP_LL_0,
+    A2DP_APTX_ADAPTIVE_TTP_LL_1,
+    A2DP_APTX_ADAPTIVE_TTP_HQ_0,
+    A2DP_APTX_ADAPTIVE_TTP_HQ_1,
+    A2DP_APTX_ADAPTIVE_TTP_TWS_0,
+    A2DP_APTX_ADAPTIVE_TTP_TWS_1,
+    A2DP_APTX_ADAPTIVE_EOC0,
+    A2DP_APTX_ADAPTIVE_EOC1,
+    A2DP_APTX_ADAPTIVE_EOC2,
+
+    //BTAV_A2DP_CODEC_BITS_PER_SAMPLE_24 /* bits_per_sample */
+    BTAV_A2DP_CODEC_BITS_PER_SAMPLE_16 /* bits_per_sample */
+};
+/* Default aptX-adaptive offload codec configuration */
+static const tA2DP_APTX_ADAPTIVE_CIE a2dp_aptx_adaptive_default_offload_config = {
+//static const tA2DP_APTX_ADAPTIVE_CIE a2dp_aptx_adaptive_default_config = {
+    A2DP_APTX_ADAPTIVE_VENDOR_ID,          /* vendorId */
+    A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH, /* codecId */
+    A2DP_APTX_ADAPTIVE_SAMPLERATE_48000,   /* sampleRate */
+    A2DP_APTX_ADAPTIVE_SOURCE_TYPE_1,
+    A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO | A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO | A2DP_APTX_ADAPTIVE_CHANNELS_TWS_MONO,      /* channelMode */
+    A2DP_APTX_ADAPTIVE_TTP_LL_0,
+    A2DP_APTX_ADAPTIVE_TTP_LL_1,
+    A2DP_APTX_ADAPTIVE_TTP_HQ_0,
+    A2DP_APTX_ADAPTIVE_TTP_HQ_1,
+    A2DP_APTX_ADAPTIVE_TTP_TWS_0,
+    A2DP_APTX_ADAPTIVE_TTP_TWS_1,
+    A2DP_APTX_ADAPTIVE_EOC0,
+    A2DP_APTX_ADAPTIVE_EOC1,
+    A2DP_APTX_ADAPTIVE_EOC2,
+
+    //BTAV_A2DP_CODEC_BITS_PER_SAMPLE_24 /* bits_per_sample */
+    BTAV_A2DP_CODEC_BITS_PER_SAMPLE_16 /* bits_per_sample */
+};
+tA2DP_APTX_ADAPTIVE_CIE a2dp_aptx_adaptive_caps, a2dp_aptx_adaptive_default_config;
+
+
+static const tA2DP_ENCODER_INTERFACE a2dp_encoder_interface_aptx_adaptive = {
+    a2dp_vendor_aptx_adaptive_encoder_init,
+    a2dp_vendor_aptx_adaptive_encoder_cleanup,
+    a2dp_vendor_aptx_adaptive_feeding_reset,
+    a2dp_vendor_aptx_adaptive_feeding_flush,
+    a2dp_vendor_aptx_adaptive_get_encoder_interval_ms, // _get_encoder_interval_ms
+    a2dp_vendor_aptx_adaptive_send_frames,
+    nullptr,  // set_transmit_queue_length
+};
+
+UNUSED_ATTR static tA2DP_STATUS A2DP_CodecInfoMatchesCapabilityAptxAdaptive(
+    const tA2DP_APTX_ADAPTIVE_CIE* p_cap, const uint8_t* p_codec_info,
+    bool is_peer_codec_info);
+
+// Builds the aptX-adaptive Media Codec Capabilities byte sequence beginning from the
+// LOSC octet. |media_type| is the media type |AVDT_MEDIA_TYPE_*|.
+// |p_ie| is a pointer to the aptX-adaptive Codec Information Element information.
+// The result is stored in |p_result|. Returns A2DP_SUCCESS on success,
+// otherwise the corresponding A2DP error status code.
+static tA2DP_STATUS A2DP_BuildInfoAptxAdaptive(uint8_t media_type,
+                                         const tA2DP_APTX_ADAPTIVE_CIE* p_ie,
+                                         uint8_t* p_result) {
+  if (p_ie == NULL || p_result == NULL) {
+    return A2DP_INVALID_PARAMS;
+  }
+
+  //*p_result++ = A2DP_APTX_ADAPTIVE_CODEC_LEN_1;
+  *p_result++ = A2DP_APTX_ADAPTIVE_CODEC_LEN;
+  *p_result++ = (media_type << 4);
+  *p_result++ = A2DP_MEDIA_CT_NON_A2DP;
+  *p_result++ = (uint8_t)(p_ie->vendorId & 0x000000FF);
+  *p_result++ = (uint8_t)((p_ie->vendorId & 0x0000FF00) >> 8);
+  *p_result++ = (uint8_t)((p_ie->vendorId & 0x00FF0000) >> 16);
+  *p_result++ = (uint8_t)((p_ie->vendorId & 0xFF000000) >> 24);
+  *p_result++ = (uint8_t)(p_ie->codecId & 0x00FF);
+  *p_result++ = (uint8_t)((p_ie->codecId & 0xFF00) >> 8);
+  *p_result++ = p_ie->sampleRate | p_ie->sourceType;
+  *p_result++ = p_ie->channelMode;
+  *p_result++ = p_ie->ttp_ll_0;
+  *p_result++ = p_ie->ttp_ll_1;
+  *p_result++ = p_ie->ttp_hq_0;
+  *p_result++ = p_ie->ttp_hq_1;
+  *p_result++ = p_ie->ttp_tws_0;
+  *p_result++ = p_ie->ttp_tws_1;
+  *p_result++ = p_ie->eoc0;
+  *p_result++ = p_ie->eoc1;
+  *p_result++ = p_ie->eoc2;
+  //*p_result++ = p_ie->ttp_tws_0;
+  //*p_result++ = p_ie->ttp_tws_1;
+
+  return A2DP_SUCCESS;
+}
+
+// Parses the aptX-adaptive Media Codec Capabilities byte sequence beginning from the
+// LOSC octet. The result is stored in |p_ie|. The byte sequence to parse is
+// |p_codec_info|. If |is_capability| is true, the byte sequence is
+// codec capabilities, otherwise is codec configuration.
+// Returns A2DP_SUCCESS on success, otherwise the corresponding A2DP error
+// status code.
+static tA2DP_STATUS A2DP_ParseInfoAptxAdaptive(tA2DP_APTX_ADAPTIVE_CIE* p_ie,
+                                         const uint8_t* p_codec_info,
+                                         bool is_capability) {
+  uint8_t losc;
+  uint8_t media_type;
+  tA2DP_CODEC_TYPE codec_type;
+
+  if (p_ie == NULL || p_codec_info == NULL) return A2DP_INVALID_PARAMS;
+
+  // Check the codec capability length
+  losc = *p_codec_info++;
+
+  if (losc != A2DP_APTX_ADAPTIVE_CODEC_LEN) {
+    LOG_ERROR(LOG_TAG, "%s: A2DP_APTX_ADAPTIVE_CODEC_LEN fail", __func__);
+    return A2DP_WRONG_CODEC;
+  }
+
+  media_type = (*p_codec_info++) >> 4;
+  codec_type = *p_codec_info++;
+  /* Check the Media Type and Media Codec Type */
+  if (media_type != AVDT_MEDIA_TYPE_AUDIO ||
+      codec_type != A2DP_MEDIA_CT_NON_A2DP) {
+    LOG_ERROR(LOG_TAG, "%s: A2DP_MEDIA_CT_NON_A2DP ID", __func__);
+    return A2DP_WRONG_CODEC;
+  }
+
+  // Check the Vendor ID and Codec ID */
+  p_ie->vendorId = (*p_codec_info & 0x000000FF) |
+                   (*(p_codec_info + 1) << 8 & 0x0000FF00) |
+                   (*(p_codec_info + 2) << 16 & 0x00FF0000) |
+                   (*(p_codec_info + 3) << 24 & 0xFF000000);
+  p_codec_info += 4;
+  p_ie->codecId =
+      (*p_codec_info & 0x00FF) | (*(p_codec_info + 1) << 8 & 0xFF00);
+  p_codec_info += 2;
+  if (p_ie->vendorId != A2DP_APTX_ADAPTIVE_VENDOR_ID ||
+      p_ie->codecId != A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH) {
+      LOG_ERROR(LOG_TAG, "%s: A2DP_APTX_ADAPTIVE ID WRONG CODEC", __func__);
+    return A2DP_WRONG_CODEC;
+  }
+
+  //p_ie->sourceType = *p_codec_info & 0x0F;
+  //p_ie->sampleRate = *p_codec_info & 0xF0;
+  p_ie->sourceType = *p_codec_info & 0x07;
+  p_ie->sampleRate = *p_codec_info & 0xF8;
+  p_codec_info++;
+
+  p_ie->channelMode = *p_codec_info & 0x0F;
+  p_codec_info++;
+  //p_ie->channelMode = *(p_codec_info ++);
+  p_ie->ttp_ll_0 = *(p_codec_info ++);
+  p_ie->ttp_ll_1 = *(p_codec_info ++);
+  p_ie->ttp_hq_0 = *(p_codec_info ++);
+  p_ie->ttp_hq_1 = *(p_codec_info ++);
+  p_ie->ttp_tws_0 = *(p_codec_info ++);
+  p_ie->ttp_tws_1 = *(p_codec_info ++);
+  p_ie->eoc0 = *(p_codec_info ++);
+  p_ie->eoc1 = *(p_codec_info ++);
+  p_ie->eoc2 = *(p_codec_info ++);
+
+  if (is_capability) return A2DP_SUCCESS;
+
+  if (A2DP_BitsSet(p_ie->sampleRate) != A2DP_SET_ONE_BIT)
+    return A2DP_BAD_SAMP_FREQ;
+  if (A2DP_BitsSet(p_ie->channelMode) != A2DP_SET_ONE_BIT)
+    return A2DP_BAD_CH_MODE;
+
+  return A2DP_SUCCESS;
+}
+
+bool A2DP_IsVendorSourceCodecValidAptxAdaptive(const uint8_t* p_codec_info) {
+  tA2DP_APTX_ADAPTIVE_CIE cfg_cie;
+
+  /* Use a liberal check when parsing the codec info */
+  return (A2DP_ParseInfoAptxAdaptive(&cfg_cie, p_codec_info, false) ==
+          A2DP_SUCCESS) ||
+         (A2DP_ParseInfoAptxAdaptive(&cfg_cie, p_codec_info, true) == A2DP_SUCCESS);
+}
+
+bool A2DP_IsVendorPeerSinkCodecValidAptxAdaptive(const uint8_t* p_codec_info) {
+  tA2DP_APTX_ADAPTIVE_CIE cfg_cie;
+
+  /* Use a liberal check when parsing the codec info */
+  return (A2DP_ParseInfoAptxAdaptive(&cfg_cie, p_codec_info, false) ==
+          A2DP_SUCCESS) ||
+         (A2DP_ParseInfoAptxAdaptive(&cfg_cie, p_codec_info, true) == A2DP_SUCCESS);
+}
+
+// Checks whether A2DP aptX-adaptive codec configuration matches with a device's
+// codec capabilities. |p_cap| is the aptX-adaptive codec configuration.
+// |p_codec_info| is the device's codec capabilities.
+// If |is_capability| is true, the byte sequence is codec capabilities,
+// otherwise is codec configuration.
+// |p_codec_info| contains the codec capabilities for a peer device that
+// is acting as an A2DP source.
+// Returns A2DP_SUCCESS if the codec configuration matches with capabilities,
+// otherwise the corresponding A2DP error status code.
+static tA2DP_STATUS A2DP_CodecInfoMatchesCapabilityAptxAdaptive(
+    const tA2DP_APTX_ADAPTIVE_CIE* p_cap, const uint8_t* p_codec_info,
+    bool is_capability) {
+  tA2DP_STATUS status;
+  tA2DP_APTX_ADAPTIVE_CIE cfg_cie;
+
+  /* parse configuration */
+  status = A2DP_ParseInfoAptxAdaptive(&cfg_cie, p_codec_info, is_capability);
+  if (status != A2DP_SUCCESS) {
+    LOG_ERROR(LOG_TAG, "%s: parsing failed %d", __func__, status);
+    return status;
+  }
+
+  /* verify that each parameter is in range */
+
+  LOG_DEBUG(LOG_TAG, "%s: FREQ peer: 0x%x, capability 0x%x", __func__,
+            cfg_cie.sampleRate, p_cap->sampleRate);
+  LOG_DEBUG(LOG_TAG, "%s: CH_MODE peer: 0x%x, capability 0x%x", __func__,
+            cfg_cie.channelMode, p_cap->channelMode);
+
+  /* sampling frequency */
+  if ((cfg_cie.sampleRate & p_cap->sampleRate) == 0) return A2DP_NS_SAMP_FREQ;
+
+  /* channel mode */
+  if ((cfg_cie.channelMode & p_cap->channelMode) == 0) return A2DP_NS_CH_MODE;
+
+  return A2DP_SUCCESS;
+}
+
+bool A2DP_VendorUsesRtpHeaderAptxAdaptive(UNUSED_ATTR bool content_protection_enabled,
+                                    UNUSED_ATTR const uint8_t* p_codec_info) {
+  return true;
+}
+
+const char* A2DP_VendorCodecNameAptxAdaptive(
+    UNUSED_ATTR const uint8_t* p_codec_info) {
+  return "aptX-adaptive";
+}
+
+bool A2DP_VendorCodecTypeEqualsAptxAdaptive(const uint8_t* p_codec_info_a,
+                                      const uint8_t* p_codec_info_b) {
+  tA2DP_APTX_ADAPTIVE_CIE aptx_adaptive_cie_a;
+  tA2DP_APTX_ADAPTIVE_CIE aptx_adaptive_cie_b;
+
+  // Check whether the codec info contains valid data
+  tA2DP_STATUS a2dp_status =
+      A2DP_ParseInfoAptxAdaptive(&aptx_adaptive_cie_a, p_codec_info_a, true);
+  if (a2dp_status != A2DP_SUCCESS) {
+    LOG_ERROR(LOG_TAG, "%s: cannot decode codec information: %d", __func__,
+              a2dp_status);
+    return false;
+  }
+  a2dp_status = A2DP_ParseInfoAptxAdaptive(&aptx_adaptive_cie_b, p_codec_info_b, true);
+  if (a2dp_status != A2DP_SUCCESS) {
+    LOG_ERROR(LOG_TAG, "%s: cannot decode codec information: %d", __func__,
+              a2dp_status);
+    return false;
+  }
+
+  return true;
+}
+
+bool A2DP_VendorCodecEqualsAptxAdaptive(const uint8_t* p_codec_info_a,
+                                  const uint8_t* p_codec_info_b) {
+  tA2DP_APTX_ADAPTIVE_CIE aptx_adaptive_cie_a;
+  tA2DP_APTX_ADAPTIVE_CIE aptx_adaptive_cie_b;
+
+  // Check whether the codec info contains valid data
+  tA2DP_STATUS a2dp_status =
+      A2DP_ParseInfoAptxAdaptive(&aptx_adaptive_cie_a, p_codec_info_a, true);
+  if (a2dp_status != A2DP_SUCCESS) {
+    LOG_ERROR(LOG_TAG, "%s: cannot decode codec information: %d", __func__,
+              a2dp_status);
+    return false;
+  }
+  a2dp_status = A2DP_ParseInfoAptxAdaptive(&aptx_adaptive_cie_b, p_codec_info_b, true);
+  if (a2dp_status != A2DP_SUCCESS) {
+    LOG_ERROR(LOG_TAG, "%s: cannot decode codec information: %d", __func__,
+              a2dp_status);
+    return false;
+  }
+
+  return (aptx_adaptive_cie_a.sampleRate == aptx_adaptive_cie_b.sampleRate) &&
+         (aptx_adaptive_cie_a.channelMode == aptx_adaptive_cie_b.channelMode);
+}
+
+int A2DP_VendorGetTrackSampleRateAptxAdaptive(const uint8_t* p_codec_info) {
+  tA2DP_APTX_ADAPTIVE_CIE aptx_adaptive_cie;
+
+  // Check whether the codec info contains valid data
+  tA2DP_STATUS a2dp_status =
+      A2DP_ParseInfoAptxAdaptive(&aptx_adaptive_cie, p_codec_info, false);
+  if (a2dp_status != A2DP_SUCCESS) {
+    LOG_ERROR(LOG_TAG, "%s: cannot decode codec information: %d", __func__,
+              a2dp_status);
+    return -1;
+  }
+
+  if (aptx_adaptive_cie.sampleRate == A2DP_APTX_ADAPTIVE_SAMPLERATE_44100) return 44100;
+  if (aptx_adaptive_cie.sampleRate == A2DP_APTX_ADAPTIVE_SAMPLERATE_48000) return 48000;
+
+  return -1;
+}
+
+int A2DP_VendorGetTrackBitsPerSampleAptxAdaptive(const uint8_t* p_codec_info) {
+  tA2DP_APTX_ADAPTIVE_CIE aptx_adaptive_cie;
+
+  // Check whether the codec info contains valid data
+  tA2DP_STATUS a2dp_status =
+      A2DP_ParseInfoAptxAdaptive(&aptx_adaptive_cie, p_codec_info, false);
+  if (a2dp_status != A2DP_SUCCESS) {
+    LOG_ERROR(LOG_TAG, "%s: cannot decode codec information: %d", __func__,
+              a2dp_status);
+    return -1;
+  }
+
+  return 16;
+}
+
+int A2DP_VendorGetTrackChannelCountAptxAdaptive(const uint8_t* p_codec_info) {
+  tA2DP_APTX_ADAPTIVE_CIE aptx_adaptive_cie;
+
+  // Check whether the codec info contains valid data
+  tA2DP_STATUS a2dp_status =
+      A2DP_ParseInfoAptxAdaptive(&aptx_adaptive_cie, p_codec_info, false);
+  if (a2dp_status != A2DP_SUCCESS) {
+    LOG_ERROR(LOG_TAG, "%s: cannot decode codec information: %d", __func__,
+              a2dp_status);
+    return -1;
+  }
+
+  switch (aptx_adaptive_cie.channelMode) {
+    case A2DP_APTX_ADAPTIVE_CHANNELS_MONO:
+      return 1;
+    case A2DP_APTX_ADAPTIVE_CHANNELS_TWS_MONO:
+      return 1;
+    case A2DP_APTX_ADAPTIVE_CHANNELS_STEREO:
+      return 2;
+    case A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO:
+      return 2;
+    case A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO:
+      return 2;
+  }
+
+  return -1;
+}
+
+bool A2DP_VendorGetPacketTimestampAptxAdaptive(
+    UNUSED_ATTR const uint8_t* p_codec_info, const uint8_t* p_data,
+    uint32_t* p_timestamp) {
+  // TODO: Is this function really codec-specific?
+  *p_timestamp = *(const uint32_t*)p_data;
+  return true;
+}
+
+bool A2DP_VendorBuildCodecHeaderAptxAdaptive(UNUSED_ATTR const uint8_t* p_codec_info,
+                                       UNUSED_ATTR BT_HDR* p_buf,
+                                       UNUSED_ATTR uint16_t frames_per_packet) {
+  // Nothing to do
+  return true;
+}
+
+bool A2DP_VendorDumpCodecInfoAptxAdaptive(const uint8_t* p_codec_info) {
+  tA2DP_STATUS a2dp_status;
+  tA2DP_APTX_ADAPTIVE_CIE aptx_adaptive_cie;
+
+  LOG_DEBUG(LOG_TAG, "%s", __func__);
+
+  a2dp_status = A2DP_ParseInfoAptxAdaptive(&aptx_adaptive_cie, p_codec_info, true);
+  if (a2dp_status != A2DP_SUCCESS) {
+    LOG_ERROR(LOG_TAG, "%s: A2DP_ParseInfoAptxAdaptive fail:%d", __func__,
+              a2dp_status);
+    return false;
+  }
+
+  LOG_DEBUG(LOG_TAG, "\tsamp_freq: 0x%x", aptx_adaptive_cie.sampleRate);
+  if (aptx_adaptive_cie.sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_44100) {
+    LOG_DEBUG(LOG_TAG, "\tsamp_freq: (44100)");
+  }
+  if (aptx_adaptive_cie.sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_48000) {
+    LOG_DEBUG(LOG_TAG, "\tsamp_freq: (48000)");
+  }
+
+  LOG_DEBUG(LOG_TAG, "\tch_mode: 0x%x", aptx_adaptive_cie.channelMode);
+  if (aptx_adaptive_cie.channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_MONO) {
+    LOG_DEBUG(LOG_TAG, "\tch_mode: (Mono)");
+  }
+
+  if (aptx_adaptive_cie.channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_MONO) {
+    LOG_DEBUG(LOG_TAG, "\tch_mode: (TWS Mono)");
+  }
+  if (aptx_adaptive_cie.channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_STEREO) {
+    LOG_DEBUG(LOG_TAG, "\tch_mode: (Stereo)");
+  }
+
+  if (aptx_adaptive_cie.channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO) {
+    LOG_DEBUG(LOG_TAG, "\tch_mode: (Joint Stereo)");
+  }
+  if (aptx_adaptive_cie.channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO) {
+    LOG_DEBUG(LOG_TAG, "\tch_mode: (TWS Stereo)");
+  }
+
+  return true;
+}
+
+const tA2DP_ENCODER_INTERFACE* A2DP_VendorGetEncoderInterfaceAptxAdaptive(
+    const uint8_t* p_codec_info) {
+  if (!A2DP_IsVendorSourceCodecValidAptxAdaptive(p_codec_info)) return NULL;
+
+  return &a2dp_encoder_interface_aptx_adaptive;
+}
+
+bool A2DP_VendorAdjustCodecAptxAdaptive(uint8_t* p_codec_info) {
+  tA2DP_APTX_ADAPTIVE_CIE cfg_cie;
+
+  // Nothing to do: just verify the codec info is valid
+  if (A2DP_ParseInfoAptxAdaptive(&cfg_cie, p_codec_info, true) != A2DP_SUCCESS)
+    return false;
+
+  return true;
+}
+
+btav_a2dp_codec_index_t A2DP_VendorSourceCodecIndexAptxAdaptive(
+    const uint8_t* p_codec_info) {
+  return BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE;
+}
+
+const char* A2DP_VendorCodecIndexStrAptxAdaptive(void) { return "aptX-adaptive"; }
+
+bool A2DP_VendorInitCodecConfigAptxAdaptive(tAVDT_CFG* p_cfg) {
+  if (A2DP_BuildInfoAptxAdaptive(AVDT_MEDIA_TYPE_AUDIO, &a2dp_aptx_adaptive_caps,
+                           p_cfg->codec_info) != A2DP_SUCCESS) {
+    return false;
+  }
+
+#if (BTA_AV_CO_CP_SCMS_T == TRUE)
+  /* Content protection info - support SCMS-T */
+  uint8_t* p = p_cfg->protect_info;
+  *p++ = AVDT_CP_LOSC;
+  UINT16_TO_STREAM(p, AVDT_CP_SCMS_T_ID);
+  p_cfg->num_protect = 1;
+#endif
+
+  return true;
+}
+
+A2dpCodecConfigAptxAdaptive::A2dpCodecConfigAptxAdaptive(
+    btav_a2dp_codec_priority_t codec_priority)
+    : A2dpCodecConfig(BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE, "aptX-adaptive",
+                      codec_priority) {
+  // Compute the local capability
+    if (A2DP_GetOffloadStatus()) {
+      a2dp_aptx_adaptive_caps = a2dp_aptx_adaptive_offload_caps;
+      a2dp_aptx_adaptive_default_config = a2dp_aptx_adaptive_default_offload_config;
+    } else {
+      a2dp_aptx_adaptive_caps = a2dp_aptx_adaptive_src_caps;
+      a2dp_aptx_adaptive_default_config = a2dp_aptx_adaptive_default_src_config;
+    }
+  if (a2dp_aptx_adaptive_caps.sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_44100) {
+    codec_local_capability_.sample_rate |= BTAV_A2DP_CODEC_SAMPLE_RATE_44100;
+  }
+  if (a2dp_aptx_adaptive_caps.sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_48000) {
+    codec_local_capability_.sample_rate |= BTAV_A2DP_CODEC_SAMPLE_RATE_48000;
+  }
+  codec_local_capability_.bits_per_sample = a2dp_aptx_adaptive_caps.bits_per_sample;
+  if (a2dp_aptx_adaptive_caps.channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_MONO) {
+    codec_local_capability_.channel_mode |= BTAV_A2DP_CODEC_CHANNEL_MODE_MONO;
+  }
+  if (a2dp_aptx_adaptive_caps.channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_MONO) {
+    codec_local_capability_.channel_mode |= BTAV_A2DP_CODEC_CHANNEL_MODE_MONO;
+  }
+  if (a2dp_aptx_adaptive_caps.channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_STEREO) {
+    codec_local_capability_.channel_mode |= BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+  }
+
+  if (a2dp_aptx_adaptive_caps.channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO) {
+    codec_local_capability_.channel_mode |= BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+  }
+  if (a2dp_aptx_adaptive_caps.channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO) {
+    codec_local_capability_.channel_mode |= BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+  }
+
+}
+
+A2dpCodecConfigAptxAdaptive::~A2dpCodecConfigAptxAdaptive() {}
+
+bool A2dpCodecConfigAptxAdaptive::init() {
+  if (!isValid()) return false;
+
+  if (A2DP_GetOffloadStatus()) {
+    if (A2DP_IsCodecEnabledInOffload(BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE)) {
+      LOG_ERROR(LOG_TAG, "%s: APTX-Adaptive enabled in offload mode", __func__);
+      return true;
+    } else {
+      LOG_ERROR(LOG_TAG, "%s: APTX-Adaptive disabled in offload mode", __func__);
+      return false;
+    }
+  }
+  // Load the encoder
+  if (!A2DP_VendorLoadEncoderAptxAdaptive()) {
+    LOG_ERROR(LOG_TAG, "%s: cannot load the encoder", __func__);
+    return false;
+  }
+
+  return true;
+}
+
+bool A2dpCodecConfigAptxAdaptive::useRtpHeaderMarkerBit() const { return false; }
+
+//
+// Selects the best sample rate from |sampleRate|.
+// The result is stored in |p_result| and p_codec_config|.
+// Returns true if a selection was made, otherwise false.
+//
+static bool select_best_sample_rate(uint8_t sampleRate,
+                                    tA2DP_APTX_ADAPTIVE_CIE* p_result,
+                                    btav_a2dp_codec_config_t* p_codec_config) {
+  if (sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_48000) {
+    p_result->sampleRate = A2DP_APTX_ADAPTIVE_SAMPLERATE_48000;
+    p_codec_config->sample_rate = BTAV_A2DP_CODEC_SAMPLE_RATE_48000;
+    return true;
+  }
+/*
+don't support 44.1 yet
+  if (sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_44100) {
+    p_result->sampleRate = A2DP_APTX_ADAPTIVE_SAMPLERATE_44100;
+    p_codec_config->sample_rate = BTAV_A2DP_CODEC_SAMPLE_RATE_44100;
+    return true;
+  }
+*/
+  return false;
+}
+
+//
+// Selects the audio sample rate from |p_codec_audio_config|.
+// |sampleRate| contains the capability.
+// The result is stored in |p_result| and |p_codec_config|.
+// Returns true if a selection was made, otherwise false.
+//
+static bool select_audio_sample_rate(
+    const btav_a2dp_codec_config_t* p_codec_audio_config, uint8_t sampleRate,
+    tA2DP_APTX_ADAPTIVE_CIE* p_result, btav_a2dp_codec_config_t* p_codec_config) {
+  switch (p_codec_audio_config->sample_rate) {
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_44100:
+      if (sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_44100) {
+        p_result->sampleRate = A2DP_APTX_ADAPTIVE_SAMPLERATE_44100;
+        p_codec_config->sample_rate = BTAV_A2DP_CODEC_SAMPLE_RATE_44100;
+        return true;
+      }
+      break;
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_48000:
+      if (sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_48000) {
+        p_result->sampleRate = A2DP_APTX_ADAPTIVE_SAMPLERATE_48000;
+        p_codec_config->sample_rate = BTAV_A2DP_CODEC_SAMPLE_RATE_48000;
+        return true;
+      }
+      break;
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_88200:
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_96000:
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_176400:
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_192000:
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_NONE:
+    default:
+      break;
+  }
+  return false;
+}
+
+//
+// Selects the best bits per sample.
+// The result is stored in |p_codec_config|.
+// Returns true if a selection was made, otherwise false.
+//
+static bool select_best_bits_per_sample(
+    btav_a2dp_codec_config_t* p_codec_config) {
+  p_codec_config->bits_per_sample = BTAV_A2DP_CODEC_BITS_PER_SAMPLE_16;
+  return true;
+}
+
+//
+// Selects the audio bits per sample from |p_codec_audio_config|.
+// The result is stored in |p_codec_config|.
+// Returns true if a selection was made, otherwise false.
+//
+static bool select_audio_bits_per_sample(
+    const btav_a2dp_codec_config_t* p_codec_audio_config,
+    btav_a2dp_codec_config_t* p_codec_config) {
+  switch (p_codec_audio_config->bits_per_sample) {
+    case BTAV_A2DP_CODEC_BITS_PER_SAMPLE_24:
+       p_codec_config->bits_per_sample = BTAV_A2DP_CODEC_BITS_PER_SAMPLE_24;
+       return true;
+    case BTAV_A2DP_CODEC_BITS_PER_SAMPLE_32:
+       p_codec_config->bits_per_sample = BTAV_A2DP_CODEC_BITS_PER_SAMPLE_32;
+       return true;
+    case BTAV_A2DP_CODEC_BITS_PER_SAMPLE_16:
+      p_codec_config->bits_per_sample = BTAV_A2DP_CODEC_BITS_PER_SAMPLE_16;
+      return true;
+    case BTAV_A2DP_CODEC_BITS_PER_SAMPLE_NONE:
+      break;
+  }
+  return false;
+}
+
+//
+// Selects the best channel mode from |channelMode|.
+// The result is stored in |p_result| and |p_codec_config|.
+// Returns true if a selection was made, otherwise false.
+//
+static bool select_best_channel_mode(uint8_t channelMode,
+                                     tA2DP_APTX_ADAPTIVE_CIE* p_result,
+                                     btav_a2dp_codec_config_t* p_codec_config) {
+  if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO) {
+    p_result->channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO;
+    p_codec_config->channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+    return true;
+  }
+
+  if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO) {
+    p_result->channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO;
+    p_codec_config->channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+    return true;
+  }
+
+  if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_STEREO) {
+    p_result->channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_STEREO;
+    p_codec_config->channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+    return true;
+  }
+  if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_MONO) {
+    p_result->channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_MONO;
+    p_codec_config->channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_MONO;
+    return true;
+  }
+  if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_MONO) {
+    p_result->channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_TWS_MONO;
+    p_codec_config->channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_MONO;
+    return true;
+  }
+  return false;
+}
+
+//
+// Selects the audio channel mode from |p_codec_audio_config|.
+// |channelMode| contains the capability.
+// The result is stored in |p_result| and |p_codec_config|.
+// Returns true if a selection was made, otherwise false.
+//
+static bool select_audio_channel_mode(
+    const btav_a2dp_codec_config_t* p_codec_audio_config, uint8_t channelMode,
+    tA2DP_APTX_ADAPTIVE_CIE* p_result, btav_a2dp_codec_config_t* p_codec_config) {
+  switch (p_codec_audio_config->channel_mode) {
+    case BTAV_A2DP_CODEC_CHANNEL_MODE_MONO:
+      if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_MONO) {
+        p_result->channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_MONO;
+        p_codec_config->channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_MONO;
+        return true;
+      }
+      if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_MONO) {
+        p_result->channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_TWS_MONO;
+        p_codec_config->channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_MONO;
+        return true;
+      }
+      break;
+    case BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO:
+      if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO) {
+        p_result->channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO;
+        p_codec_config->channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+        return true;
+      }
+      if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO) {
+        p_result->channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO;
+        p_codec_config->channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+        return true;
+      }
+      if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_STEREO) {
+        p_result->channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_STEREO;
+        p_codec_config->channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+        return true;
+      }
+      break;
+    case BTAV_A2DP_CODEC_CHANNEL_MODE_NONE:
+      break;
+  }
+
+  return false;
+}
+
+bool A2dpCodecConfigAptxAdaptive::setCodecConfig(const uint8_t* p_peer_codec_info,
+                                           bool is_capability,
+                                           uint8_t* p_result_codec_config) {
+  std::lock_guard<std::recursive_mutex> lock(codec_mutex_);
+  tA2DP_APTX_ADAPTIVE_CIE sink_info_cie;
+  tA2DP_APTX_ADAPTIVE_CIE result_config_cie;
+  uint8_t sampleRate;
+  uint8_t channelMode;
+
+  // Save the internal state
+  btav_a2dp_codec_config_t saved_codec_config = codec_config_;
+  btav_a2dp_codec_config_t saved_codec_capability = codec_capability_;
+  btav_a2dp_codec_config_t saved_codec_selectable_capability =
+      codec_selectable_capability_;
+  btav_a2dp_codec_config_t saved_codec_user_config = codec_user_config_;
+  btav_a2dp_codec_config_t saved_codec_audio_config = codec_audio_config_;
+  uint8_t saved_ota_codec_config[AVDT_CODEC_SIZE];
+  uint8_t saved_ota_codec_peer_capability[AVDT_CODEC_SIZE];
+  uint8_t saved_ota_codec_peer_config[AVDT_CODEC_SIZE];
+  memcpy(saved_ota_codec_config, ota_codec_config_, sizeof(ota_codec_config_));
+  memcpy(saved_ota_codec_peer_capability, ota_codec_peer_capability_,
+         sizeof(ota_codec_peer_capability_));
+  memcpy(saved_ota_codec_peer_config, ota_codec_peer_config_,
+         sizeof(ota_codec_peer_config_));
+
+  tA2DP_STATUS status =
+      A2DP_ParseInfoAptxAdaptive(&sink_info_cie, p_peer_codec_info, is_capability);
+  if (status != A2DP_SUCCESS) {
+    LOG_ERROR(LOG_TAG, "%s: can't parse peer's Sink capabilities: error = %d",
+              __func__, status);
+    goto fail;
+  }
+
+  //
+  // Build the preferred configuration
+  //
+  memset(&result_config_cie, 0, sizeof(result_config_cie));
+  result_config_cie.vendorId = a2dp_aptx_adaptive_caps.vendorId;
+  result_config_cie.codecId = a2dp_aptx_adaptive_caps.codecId;
+
+  //
+  // Select the sample frequency
+  //
+  sampleRate = a2dp_aptx_adaptive_caps.sampleRate & sink_info_cie.sampleRate;
+  codec_config_.sample_rate = BTAV_A2DP_CODEC_SAMPLE_RATE_NONE;
+  switch (codec_user_config_.sample_rate) {
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_44100:
+      if (sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_44100) {
+        result_config_cie.sampleRate = A2DP_APTX_ADAPTIVE_SAMPLERATE_44100;
+        codec_capability_.sample_rate = codec_user_config_.sample_rate;
+        codec_config_.sample_rate = codec_user_config_.sample_rate;
+      }
+      break;
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_48000:
+      if (sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_48000) {
+        result_config_cie.sampleRate = A2DP_APTX_ADAPTIVE_SAMPLERATE_48000;
+        codec_capability_.sample_rate = codec_user_config_.sample_rate;
+        codec_config_.sample_rate = codec_user_config_.sample_rate;
+      }
+      break;
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_88200:
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_96000:
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_176400:
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_192000:
+    case BTAV_A2DP_CODEC_SAMPLE_RATE_NONE:
+    default:
+      codec_capability_.sample_rate = BTAV_A2DP_CODEC_SAMPLE_RATE_NONE;
+      codec_config_.sample_rate = BTAV_A2DP_CODEC_SAMPLE_RATE_NONE;
+      break;
+  }
+
+  // Select the sample frequency if there is no user preference
+  do {
+    // Compute the selectable capability
+    if (sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_44100) {
+      codec_selectable_capability_.sample_rate |=
+          BTAV_A2DP_CODEC_SAMPLE_RATE_44100;
+    }
+    if (sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_48000) {
+      codec_selectable_capability_.sample_rate |=
+          BTAV_A2DP_CODEC_SAMPLE_RATE_48000;
+    }
+
+    if (codec_config_.sample_rate != BTAV_A2DP_CODEC_SAMPLE_RATE_NONE) break;
+
+    // Compute the common capability
+    if (sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_44100)
+      codec_capability_.sample_rate |= BTAV_A2DP_CODEC_SAMPLE_RATE_44100;
+    if (sampleRate & A2DP_APTX_ADAPTIVE_SAMPLERATE_48000)
+      codec_capability_.sample_rate |= BTAV_A2DP_CODEC_SAMPLE_RATE_48000;
+
+    // No user preference - try the codec audio config
+    if (select_audio_sample_rate(&codec_audio_config_, sampleRate,
+                                 &result_config_cie, &codec_config_)) {
+      break;
+    }
+
+    // No user preference - try the default config
+    if (select_best_sample_rate(
+            a2dp_aptx_adaptive_default_config.sampleRate & sink_info_cie.sampleRate,
+            &result_config_cie, &codec_config_)) {
+      break;
+    }
+
+    // No user preference - use the best match
+    if (select_best_sample_rate(sampleRate, &result_config_cie,
+                                &codec_config_)) {
+      break;
+    }
+  } while (false);
+  if (codec_config_.sample_rate == BTAV_A2DP_CODEC_SAMPLE_RATE_NONE) {
+    LOG_ERROR(LOG_TAG,
+              "%s: cannot match sample frequency: source caps = 0x%x "
+              "sink info = 0x%x",
+              __func__, a2dp_aptx_adaptive_caps.sampleRate, sink_info_cie.sampleRate);
+    goto fail;
+  }
+
+
+  //
+  // Select the bits per sample
+  //
+  // NOTE: this information is NOT included in the aptX-adaptive A2DP codec
+  // description that is sent OTA.
+  codec_config_.bits_per_sample = BTAV_A2DP_CODEC_BITS_PER_SAMPLE_NONE;
+  switch (codec_user_config_.bits_per_sample) {
+    case BTAV_A2DP_CODEC_BITS_PER_SAMPLE_16:
+      codec_capability_.bits_per_sample = codec_user_config_.bits_per_sample;
+      codec_config_.bits_per_sample = codec_user_config_.bits_per_sample;
+      break;
+
+    case BTAV_A2DP_CODEC_BITS_PER_SAMPLE_24:
+      //codec_capability_.bits_per_sample = codec_user_config_.bits_per_sample;
+      //codec_config_.bits_per_sample = codec_user_config_.bits_per_sample;
+     // break;
+    case BTAV_A2DP_CODEC_BITS_PER_SAMPLE_32:
+
+    case BTAV_A2DP_CODEC_BITS_PER_SAMPLE_NONE:
+      codec_capability_.bits_per_sample = BTAV_A2DP_CODEC_BITS_PER_SAMPLE_NONE;
+      codec_config_.bits_per_sample = BTAV_A2DP_CODEC_BITS_PER_SAMPLE_NONE;
+      break;
+  }
+  // Select the bits per sample if there is no user preference
+  do {
+    // Compute the selectable capability
+    codec_selectable_capability_.bits_per_sample =
+        a2dp_aptx_adaptive_caps.bits_per_sample;
+
+    if (codec_config_.bits_per_sample != BTAV_A2DP_CODEC_BITS_PER_SAMPLE_NONE)
+      break;
+
+    // Compute the common capability
+    //codec_capability_.bits_per_sample = BTAV_A2DP_CODEC_BITS_PER_SAMPLE_24;
+    codec_capability_.bits_per_sample = BTAV_A2DP_CODEC_BITS_PER_SAMPLE_16;
+
+    // No user preference - try the codec audio config
+    if (select_audio_bits_per_sample(&codec_audio_config_, &codec_config_)) {
+      break;
+    }
+
+    // No user preference - try the default config
+    if (select_best_bits_per_sample(&codec_config_)) {
+      break;
+    }
+
+    // No user preference - use the best match
+    // NOTE: no-op - kept here for consistency
+    if (select_best_bits_per_sample(&codec_config_)) {
+      break;
+    }
+  } while (false);
+  if (codec_config_.bits_per_sample == BTAV_A2DP_CODEC_BITS_PER_SAMPLE_NONE) {
+    LOG_ERROR(LOG_TAG,
+              "%s: cannot match bits per sample: user preference = 0x%x",
+              __func__, codec_user_config_.bits_per_sample);
+    goto fail;
+  }
+
+  //
+  // Select the channel mode
+  //
+  channelMode = a2dp_aptx_adaptive_caps.channelMode & sink_info_cie.channelMode;
+  codec_config_.channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_NONE;
+  switch (codec_user_config_.channel_mode) {
+    case BTAV_A2DP_CODEC_CHANNEL_MODE_MONO:
+      if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_MONO) {
+        result_config_cie.channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_MONO;
+        codec_capability_.channel_mode = codec_user_config_.channel_mode;
+        codec_config_.channel_mode = codec_user_config_.channel_mode;
+        break;
+      }
+      if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_MONO) {
+        result_config_cie.channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_MONO;
+        codec_capability_.channel_mode = codec_user_config_.channel_mode;
+        codec_config_.channel_mode = codec_user_config_.channel_mode;
+        break;
+      }
+      break;
+    case BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO:
+      if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO) {
+        result_config_cie.channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO;
+        codec_capability_.channel_mode = codec_user_config_.channel_mode;
+        codec_config_.channel_mode = codec_user_config_.channel_mode;
+        break;
+      }
+      if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO) {
+        result_config_cie.channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO;
+        codec_capability_.channel_mode = codec_user_config_.channel_mode;
+        codec_config_.channel_mode = codec_user_config_.channel_mode;
+        break;
+      }
+      if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_STEREO) {
+        result_config_cie.channelMode = A2DP_APTX_ADAPTIVE_CHANNELS_STEREO;
+        codec_capability_.channel_mode = codec_user_config_.channel_mode;
+        codec_config_.channel_mode = codec_user_config_.channel_mode;
+        break;
+      }
+      break;
+    case BTAV_A2DP_CODEC_CHANNEL_MODE_NONE:
+      codec_capability_.channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_NONE;
+      codec_config_.channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_NONE;
+      break;
+  }
+  // Select the channel mode if there is no user preference
+  do {
+    // Compute the selectable capability
+    if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_MONO) {
+      codec_selectable_capability_.channel_mode |=
+          BTAV_A2DP_CODEC_CHANNEL_MODE_MONO;
+    }
+    if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_MONO) {
+      codec_selectable_capability_.channel_mode |=
+          BTAV_A2DP_CODEC_CHANNEL_MODE_MONO;
+    }
+    if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO) {
+      codec_selectable_capability_.channel_mode |=
+          BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+    }
+    if  (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO) {
+      codec_selectable_capability_.channel_mode |=
+          BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+    }
+    if  (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_STEREO) {
+      codec_selectable_capability_.channel_mode |=
+          BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+    }
+
+    if (codec_config_.channel_mode != BTAV_A2DP_CODEC_CHANNEL_MODE_NONE) break;
+
+    // Compute the common capability
+    if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_MONO)
+      codec_capability_.channel_mode |= BTAV_A2DP_CODEC_CHANNEL_MODE_MONO;
+    if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_MONO)
+      codec_capability_.channel_mode |= BTAV_A2DP_CODEC_CHANNEL_MODE_MONO;
+    if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_STEREO)
+      codec_capability_.channel_mode |= BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+    if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO)
+      codec_capability_.channel_mode |= BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+    if (channelMode & A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO)
+      codec_capability_.channel_mode |= BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO;
+
+    // No user preference - try the codec audio config
+    if (select_audio_channel_mode(&codec_audio_config_, channelMode,
+                                  &result_config_cie, &codec_config_)) {
+      break;
+    }
+
+    // No user preference - try the default config
+    if (select_best_channel_mode(
+            a2dp_aptx_adaptive_default_config.channelMode & sink_info_cie.channelMode,
+            &result_config_cie, &codec_config_)) {
+      break;
+    }
+
+    // No user preference - use the best match
+    if (select_best_channel_mode(channelMode, &result_config_cie,
+                                 &codec_config_)) {
+      break;
+    }
+  } while (false);
+  if (codec_config_.channel_mode == BTAV_A2DP_CODEC_CHANNEL_MODE_NONE) {
+    LOG_ERROR(LOG_TAG,
+              "%s: cannot match channel mode: source caps = 0x%x "
+              "sink info = 0x%x",
+              __func__, a2dp_aptx_adaptive_caps.channelMode,
+              sink_info_cie.channelMode);
+    goto fail;
+  }
+
+  result_config_cie.sourceType = a2dp_aptx_adaptive_caps.sourceType;
+  result_config_cie.ttp_ll_0 = a2dp_aptx_adaptive_caps.ttp_ll_0;
+  result_config_cie.ttp_ll_1 = a2dp_aptx_adaptive_caps.ttp_ll_1;
+  result_config_cie.ttp_hq_0 = a2dp_aptx_adaptive_caps.ttp_hq_0;
+  result_config_cie.ttp_hq_1 = a2dp_aptx_adaptive_caps.ttp_hq_1;
+  result_config_cie.ttp_tws_0 = a2dp_aptx_adaptive_caps.ttp_tws_0;
+  result_config_cie.ttp_tws_1 = a2dp_aptx_adaptive_caps.ttp_tws_1;
+  result_config_cie.eoc0 = a2dp_aptx_adaptive_caps.eoc0;
+  result_config_cie.eoc1 = a2dp_aptx_adaptive_caps.eoc1;
+  result_config_cie.eoc2 = a2dp_aptx_adaptive_caps.eoc2;
+
+  if (A2DP_BuildInfoAptxAdaptive(AVDT_MEDIA_TYPE_AUDIO, &result_config_cie,
+                           p_result_codec_config) != A2DP_SUCCESS) {
+    goto fail;
+  }
+
+  if (codec_user_config_.codec_specific_2 != codec_config_.codec_specific_2 ) {
+    codec_config_.codec_specific_2 = codec_user_config_.codec_specific_2;
+  }
+  if (codec_user_config_.codec_specific_3 != codec_config_.codec_specific_3 ) {
+    codec_user_config_.codec_specific_3 = codec_config_.codec_specific_3;
+  }
+
+  LOG_ERROR(LOG_TAG, "setting codec_config_.codec_specific_4 to channelMode: %d", channelMode);
+  // Store the channel mode in spare field
+  codec_config_.codec_specific_4 &= (int64_t) CHANNEL_MODE_BACK_CHANNEL_MASK;
+  codec_config_.codec_specific_4 |= channelMode;
+
+  // Create a local copy of the peer codec capability/config, and the
+  // result codec config.
+  if (is_capability) {
+    status = A2DP_BuildInfoAptxAdaptive(AVDT_MEDIA_TYPE_AUDIO, &sink_info_cie,
+                                  ota_codec_peer_capability_);
+  } else {
+    status = A2DP_BuildInfoAptxAdaptive(AVDT_MEDIA_TYPE_AUDIO, &sink_info_cie,
+                                  ota_codec_peer_config_);
+  }
+  CHECK(status == A2DP_SUCCESS);
+  status = A2DP_BuildInfoAptxAdaptive(AVDT_MEDIA_TYPE_AUDIO, &result_config_cie,
+                                ota_codec_config_);
+  CHECK(status == A2DP_SUCCESS);
+  return true;
+
+fail:
+  // Restore the internal state
+  codec_config_ = saved_codec_config;
+  codec_capability_ = saved_codec_capability;
+  codec_selectable_capability_ = saved_codec_selectable_capability;
+  codec_user_config_ = saved_codec_user_config;
+  codec_audio_config_ = saved_codec_audio_config;
+  memcpy(ota_codec_config_, saved_ota_codec_config, sizeof(ota_codec_config_));
+  memcpy(ota_codec_peer_capability_, saved_ota_codec_peer_capability,
+         sizeof(ota_codec_peer_capability_));
+  memcpy(ota_codec_peer_config_, saved_ota_codec_peer_config,
+         sizeof(ota_codec_peer_config_));
+  return false;
+}
diff --git a/stack/a2dp/a2dp_vendor_aptx_adaptive_encoder.cc b/stack/a2dp/a2dp_vendor_aptx_adaptive_encoder.cc
new file mode 100644
index 0000000..8833633
--- /dev/null
+++ b/stack/a2dp/a2dp_vendor_aptx_adaptive_encoder.cc
@@ -0,0 +1,227 @@
+/******************************************************************************
+    Copyright (c) 2018, The Linux Foundation. All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted (subject to the limitations in the
+    disclaimer below) provided that the following conditions are met:
+
+       * Redistributions of source code must retain the above copyright
+         notice, this list of conditions and the following disclaimer.
+
+       * Redistributions in binary form must reproduce the above
+         copyright notice, this list of conditions and the following
+         disclaimer in the documentation and/or other materials provided
+         with the distribution.
+
+       * Neither the name of The Linux Foundation nor the names of its
+         contributors may be used to endorse or promote products derived
+         from this software without specific prior written permission.
+
+    NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE
+    GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT
+    HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
+    WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+    MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+    IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+    ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+    GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+    IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+    OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+    IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+#define LOG_TAG "a2dp_vendor_aptx_adaptive_encoder"
+
+#include "a2dp_vendor_aptx_adaptive_encoder.h"
+
+#include <dlfcn.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <string.h>
+#include <cmath>
+
+#include "a2dp_vendor.h"
+#include "a2dp_vendor_aptx_adaptive.h"
+#include "bt_common.h"
+#include "osi/include/log.h"
+
+#include <time.h>
+
+typedef struct {
+  uint64_t session_start_us;
+
+  size_t media_read_total_expected_packets;
+  size_t media_read_total_expected_reads_count;
+  size_t media_read_total_expected_read_bytes;
+
+  size_t media_read_total_dropped_packets;
+  size_t media_read_total_actual_reads_count;
+  size_t media_read_total_actual_read_bytes;
+} a2dp_aptx_adaptive_encoder_stats_t;
+
+
+
+typedef struct {
+  a2dp_source_read_callback_t read_callback;
+  a2dp_source_enqueue_callback_t enqueue_callback;
+
+  bool use_SCMS_T;
+  bool is_peer_edr;          // True if the peer device supports EDR
+  bool peer_supports_3mbps;  // True if the peer device supports 3Mbps EDR
+  uint16_t peer_mtu;         // // MTU of the A2DP peer
+  uint32_t timestamp;        // Timestamp for the A2DP frames
+
+  a2dp_aptx_adaptive_encoder_stats_t stats;
+
+} tA2DP_APTX_ADAPTIVE_ENCODER_CB;
+
+static tA2DP_APTX_ADAPTIVE_ENCODER_CB a2dp_aptx_adaptive_encoder_cb;
+
+
+bool A2DP_VendorLoadEncoderAptxAdaptive(void) {
+  if (A2DP_GetOffloadStatus()) {
+    LOG_INFO(LOG_TAG,"aptX-Adaptive is running in offload mode");
+    return true;
+  }
+
+
+  return true;
+}
+
+void A2DP_VendorUnloadEncoderAptxAdaptive(void) {
+  if (A2DP_GetOffloadStatus()) {
+    LOG_INFO(LOG_TAG,"aptX-Adaptive is running in offload mode");
+    return;
+  }
+}
+
+void a2dp_vendor_aptx_adaptive_encoder_init(
+    const tA2DP_ENCODER_INIT_PEER_PARAMS* p_peer_params,
+    A2dpCodecConfig* a2dp_codec_config,
+    a2dp_source_read_callback_t read_callback,
+    a2dp_source_enqueue_callback_t enqueue_callback) {
+  if (A2DP_GetOffloadStatus()) {
+    LOG_INFO(LOG_TAG,"aptX-Adaptive is running in offload mode");
+    return;
+  }
+}
+
+
+
+bool A2dpCodecConfigAptxAdaptive::updateEncoderUserConfig(
+    const tA2DP_ENCODER_INIT_PEER_PARAMS* p_peer_params, bool* p_restart_input,
+    bool* p_restart_output, bool* p_config_updated) {
+
+  btav_a2dp_codec_config_t codec_config = getCodecConfig();
+
+  a2dp_aptx_adaptive_encoder_cb.is_peer_edr = p_peer_params->is_peer_edr;
+  a2dp_aptx_adaptive_encoder_cb.peer_supports_3mbps =
+      p_peer_params->peer_supports_3mbps;
+  a2dp_aptx_adaptive_encoder_cb.peer_mtu = p_peer_params->peer_mtu;
+
+  if (a2dp_aptx_adaptive_encoder_cb.peer_mtu == 0) {
+    LOG_ERROR(LOG_TAG,
+              "%s: Cannot update the codec encoder for %s: "
+              "invalid peer MTU",
+              __func__, name().c_str());
+    return false;
+  }
+
+
+   if (!isCodecConfigEmpty(previous_codec_config)) {
+      if (previous_codec_config.codec_priority == codec_config.codec_priority &&
+          previous_codec_config.sample_rate == codec_config.sample_rate &&
+          previous_codec_config.bits_per_sample == codec_config.bits_per_sample &&
+          previous_codec_config.channel_mode == codec_config.channel_mode) {
+
+        if (previous_codec_config.codec_specific_4 != codec_config.codec_specific_4 ) {
+          // this is just our private config that has changed, no need to send
+          // a BTIF_AV_SOURCE_CONFIG_UPDATED_EVT, which restarts split audio
+          *p_config_updated = false;
+        }
+      }
+    }
+
+  previous_codec_config = codec_config;
+
+
+  return true;
+}
+
+
+
+
+
+
+void a2dp_vendor_aptx_adaptive_feeding_reset(void) {
+  if (A2DP_GetOffloadStatus()) {
+    LOG_INFO(LOG_TAG,"a2dp_vendor_aptx_adaptive_send_frames"
+                                  "aptX-Adaptive is running in offload mode");
+    return;
+  }
+}
+
+
+void a2dp_vendor_aptx_adaptive_feeding_flush(void) {
+  if (A2DP_GetOffloadStatus()) {
+    LOG_INFO(LOG_TAG,"a2dp_vendor_aptx_adaptive_send_frames"
+                                  "aptX-Adaptive is running in offload mode");
+    return;
+  }
+
+
+}
+
+period_ms_t a2dp_vendor_aptx_adaptive_get_encoder_interval_ms(void) {
+  return 0;
+}
+
+
+void a2dp_vendor_aptx_adaptive_send_frames(uint64_t timestamp_us) {
+  if (A2DP_GetOffloadStatus()) {
+    LOG_INFO(LOG_TAG,"a2dp_vendor_aptx_adaptive_send_frames"
+        "aptX-Adaptive is running in offload mode");
+    return;
+  }
+}
+
+void a2dp_vendor_aptx_adaptive_encoder_cleanup(void) {
+  //osi_free(a2dp_aptx_adaptive_encoder_cb.aptx_adaptive_encoder_state);
+  memset(&a2dp_aptx_adaptive_encoder_cb, 0, sizeof(a2dp_aptx_adaptive_encoder_cb));
+}
+
+
+period_ms_t A2dpCodecConfigAptxAdaptive::encoderIntervalMs() const {
+  return 0;
+}
+
+
+void A2dpCodecConfigAptxAdaptive::debug_codec_dump(int fd) {
+  a2dp_aptx_adaptive_encoder_stats_t* stats = &a2dp_aptx_adaptive_encoder_cb.stats;
+
+  A2dpCodecConfig::debug_codec_dump(fd);
+
+  dprintf(fd,
+          "  Packet counts (expected/dropped)                        : %zu / "
+          "%zu\n",
+          stats->media_read_total_expected_packets,
+          stats->media_read_total_dropped_packets);
+
+  dprintf(fd,
+          "  PCM read counts (expected/actual)                       : %zu / "
+          "%zu\n",
+          stats->media_read_total_expected_reads_count,
+          stats->media_read_total_actual_reads_count);
+
+  dprintf(fd,
+          "  PCM read bytes (expected/actual)                        : %zu / "
+          "%zu\n",
+          stats->media_read_total_expected_read_bytes,
+          stats->media_read_total_actual_read_bytes);
+
+  /*dprintf(fd,
+          "  Connection count                                        : %u\n",
+		  a2dp_aptx_adaptive_encoder_cb.enc_params.connection_count);*/
+}
diff --git a/stack/include/a2dp_vendor_aptx_adaptive.h b/stack/include/a2dp_vendor_aptx_adaptive.h
new file mode 100644
index 0000000..87f2513
--- /dev/null
+++ b/stack/include/a2dp_vendor_aptx_adaptive.h
@@ -0,0 +1,173 @@
+/******************************************************************************
+    Copyright (c) 2018, The Linux Foundation. All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted (subject to the limitations in the
+    disclaimer below) provided that the following conditions are met:
+
+       * Redistributions of source code must retain the above copyright
+         notice, this list of conditions and the following disclaimer.
+
+       * Redistributions in binary form must reproduce the above
+         copyright notice, this list of conditions and the following
+         disclaimer in the documentation and/or other materials provided
+         with the distribution.
+
+       * Neither the name of The Linux Foundation nor the names of its
+         contributors may be used to endorse or promote products derived
+         from this software without specific prior written permission.
+
+    NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE
+    GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT
+    HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
+    WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+    MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+    IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+    ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+    GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+    IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+    OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+    IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+//
+// A2DP Codec API for aptX-adaptive
+//
+
+#ifndef A2DP_VENDOR_APTX_ADAPTIVE_H
+#define A2DP_VENDOR_APTX_ADAPTIVE_H
+
+#include "a2dp_codec_api.h"
+#include "a2dp_vendor_aptx_adaptive_constants.h"
+#include "avdt_api.h"
+
+class A2dpCodecConfigAptxAdaptive: public A2dpCodecConfig {
+ public:
+  A2dpCodecConfigAptxAdaptive(btav_a2dp_codec_priority_t codec_priority);
+  virtual ~A2dpCodecConfigAptxAdaptive();
+
+  bool init() override;
+  period_ms_t encoderIntervalMs() const override;
+  bool setCodecConfig(const uint8_t* p_peer_codec_info, bool is_capability,
+                      uint8_t* p_result_codec_config) override;
+
+ private:
+  bool useRtpHeaderMarkerBit() const override;
+  bool updateEncoderUserConfig(
+      const tA2DP_ENCODER_INIT_PEER_PARAMS* p_peer_params,
+      bool* p_restart_input, bool* p_restart_output,
+      bool* p_config_updated) override;
+
+  btav_a2dp_codec_config_t previous_codec_config;
+
+  void debug_codec_dump(int fd) override;
+};
+
+// Checks whether the codec capabilities contain a valid A2DP aptX-adaptive Source
+// codec.
+// NOTE: only codecs that are implemented are considered valid.
+// Returns true if |p_codec_info| contains information about a valid aptX-adaptive
+// codec, otherwise false.
+bool A2DP_IsVendorSourceCodecValidAptxAdaptive(const uint8_t* p_codec_info);
+
+// Checks whether the codec capabilities contain a valid peer A2DP aptX-adaptive Sink
+// codec.
+// NOTE: only codecs that are implemented are considered valid.
+// Returns true if |p_codec_info| contains information about a valid aptX-adaptive
+// codec, otherwise false.
+bool A2DP_IsVendorPeerSinkCodecValidAptxAdaptive(const uint8_t* p_codec_info);
+
+// Checks whether the A2DP data packets should contain RTP header.
+// |content_protection_enabled| is true if Content Protection is
+// enabled. |p_codec_info| contains information about the codec capabilities.
+// Returns true if the A2DP data packets should contain RTP header, otherwise
+// false.
+bool A2DP_VendorUsesRtpHeaderAptxAdaptive(bool content_protection_enabled,
+                                    const uint8_t* p_codec_info);
+
+// Gets the A2DP aptX-adaptive codec name for a given |p_codec_info|.
+const char* A2DP_VendorCodecNameAptxAdaptive(const uint8_t* p_codec_info);
+
+// Checks whether two A2DP aptX-adaptive codecs |p_codec_info_a| and |p_codec_info_b|
+// have the same type.
+// Returns true if the two codecs have the same type, otherwise false.
+bool A2DP_VendorCodecTypeEqualsAptxAdaptive(const uint8_t* p_codec_info_a,
+                                      const uint8_t* p_codec_info_b);
+
+// Checks whether two A2DP aptX-adaptive codecs |p_codec_info_a| and |p_codec_info_b|
+// are exactly the same.
+// Returns true if the two codecs are exactly the same, otherwise false.
+// If the codec type is not aptX-adaptive, the return value is false.
+bool A2DP_VendorCodecEqualsAptxAdaptive(const uint8_t* p_codec_info_a,
+                                  const uint8_t* p_codec_info_b);
+
+// Gets the track sample rate value for the A2DP aptX-adaptive codec.
+// |p_codec_info| is a pointer to the aptX-adaptive codec_info to decode.
+// Returns the track sample rate on success, or -1 if |p_codec_info|
+// contains invalid codec information.
+int A2DP_VendorGetTrackSampleRateAptxAdaptive(const uint8_t* p_codec_info);
+
+// Gets the bits per audio sample for the A2DP aptX-adaptive codec.
+// |p_codec_info| is a pointer to the aptX-adaptive codec_info to decode.
+// Returns the bits per audio sample on success, or -1 if |p_codec_info|
+// contains invalid codec information.
+int A2DP_VendorGetTrackBitsPerSampleAptxAdaptive(const uint8_t* p_codec_info);
+
+// Gets the channel count for the A2DP aptX-adaptive codec.
+// |p_codec_info| is a pointer to the aptX-adaptive codec_info to decode.
+// Returns the channel count on success, or -1 if |p_codec_info|
+// contains invalid codec information.
+int A2DP_VendorGetTrackChannelCountAptxAdaptive(const uint8_t* p_codec_info);
+
+// Gets the A2DP aptX-adaptive audio data timestamp from an audio packet.
+// |p_codec_info| contains the codec information.
+// |p_data| contains the audio data.
+// The timestamp is stored in |p_timestamp|.
+// Returns true on success, otherwise false.
+bool A2DP_VendorGetPacketTimestampAptxAdaptive(const uint8_t* p_codec_info,
+                                         const uint8_t* p_data,
+                                         uint32_t* p_timestamp);
+
+// Builds A2DP aptX-adaptive codec header for audio data.
+// |p_codec_info| contains the codec information.
+// |p_buf| contains the audio data.
+// |frames_per_packet| is the number of frames in this packet.
+// Returns true on success, otherwise false.
+bool A2DP_VendorBuildCodecHeaderAptxAdaptive(const uint8_t* p_codec_info,
+                                       BT_HDR* p_buf,
+                                       uint16_t frames_per_packet);
+
+// Decodes and displays aptX-adaptive codec info (for debugging).
+// |p_codec_info| is a pointer to the aptX-adaptive codec_info to decode and display.
+bool A2DP_VendorDumpCodecInfoAptxAdaptive(const uint8_t* p_codec_info);
+
+// Gets the A2DP aptX-adaptive encoder interface that can be used to encode and
+// prepare A2DP packets for transmission - see |tA2DP_ENCODER_INTERFACE|.
+// |p_codec_info| contains the codec information.
+// Returns the A2DP aptX-adaptive encoder interface if the |p_codec_info| is valid
+// and supported, otherwise NULL.
+const tA2DP_ENCODER_INTERFACE* A2DP_VendorGetEncoderInterfaceAptxAdaptive(
+    const uint8_t* p_codec_info);
+
+// Adjusts the A2DP aptX-adaptive codec, based on local support and Bluetooth
+// specification.
+// |p_codec_info| contains the codec information to adjust.
+// Returns true if |p_codec_info| is valid and supported, otherwise false.
+bool A2DP_VendorAdjustCodecAptxAdaptive(uint8_t* p_codec_info);
+
+// Gets the A2DP aptX-adaptive Source codec index for a given |p_codec_info|.
+// Returns the corresponding |btav_a2dp_codec_index_t| on success,
+// otherwise |BTAV_A2DP_CODEC_INDEX_MAX|.
+btav_a2dp_codec_index_t A2DP_VendorSourceCodecIndexAptxAdaptive(
+    const uint8_t* p_codec_info);
+
+// Gets the A2DP aptX-adaptive Source codec name.
+const char* A2DP_VendorCodecIndexStrAptxAdaptive(void);
+
+// Initializes A2DP aptX-adaptive Source codec information into |tAVDT_CFG|
+// configuration entry pointed by |p_cfg|.
+bool A2DP_VendorInitCodecConfigAptxAdaptive(tAVDT_CFG* p_cfg);
+
+#endif  // A2DP_VENDOR_APTX_ADAPTIVE_H
diff --git a/stack/include/a2dp_vendor_aptx_adaptive_constants.h b/stack/include/a2dp_vendor_aptx_adaptive_constants.h
new file mode 100644
index 0000000..8933cce
--- /dev/null
+++ b/stack/include/a2dp_vendor_aptx_adaptive_constants.h
@@ -0,0 +1,89 @@
+/******************************************************************************
+    Copyright (c) 2018, The Linux Foundation. All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted (subject to the limitations in the
+    disclaimer below) provided that the following conditions are met:
+
+       * Redistributions of source code must retain the above copyright
+         notice, this list of conditions and the following disclaimer.
+
+       * Redistributions in binary form must reproduce the above
+         copyright notice, this list of conditions and the following
+         disclaimer in the documentation and/or other materials provided
+         with the distribution.
+
+       * Neither the name of The Linux Foundation nor the names of its
+         contributors may be used to endorse or promote products derived
+         from this software without specific prior written permission.
+
+    NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE
+    GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT
+    HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
+    WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+    MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+    IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+    ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+    GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+    IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+    OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+    IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+//
+// A2DP constants for aptX-adaptive codec
+//
+
+#ifndef A2DP_VENDOR_APTX_ADAPTIVE_CONSTANTS_H
+#define A2DP_VENDOR_APTX_ADAPTIVE_CONSTANTS_H
+
+/* aptX-adaptive codec specific settings */
+#define A2DP_APTX_ADAPTIVE_CODEC_LEN 19
+#define A2DP_APTX_ADAPTIVE_CODEC_LEN_1 17
+
+// actual id's ..
+#define A2DP_APTX_ADAPTIVE_VENDOR_ID               (0x000000D7)
+#define A2DP_APTX_ADAPTIVE_CODEC_ID_BLUETOOTH      (0x00AD)
+
+#define A2DP_APTX_ADAPTIVE_SAMPLERATE_44100        (0x08)
+#define A2DP_APTX_ADAPTIVE_SAMPLERATE_48000        (0x10)
+#define A2DP_APTX_ADAPTIVE_SAMPLERATE_88000        (0x20)
+#define A2DP_APTX_ADAPTIVE_SAMPLERATE_192000       (0x40)
+
+#define A2DP_APTX_ADAPTIVE_SOURCE_TYPE_1           (0x00) //(0x00)
+#define A2DP_APTX_ADAPTIVE_SOURCE_TYPE_2           (0x02) //(0x02)
+#define A2DP_APTX_ADAPTIVE_SOURCE_TYPE_3           (0x04) //(0x04)
+#define A2DP_APTX_ADAPTIVE_SOURCE_TYPE_4           (0x06) //(0x06)
+/*
+#define A2DP_APTX_ADAPTIVE_AVRCP_CUSTOM_UNSUPPORTED(0x00)
+#define A2DP_APTX_ADAPTIVE_AVRCP_CUSTOM_SUPPORTED  (0x01)
+#define A2DP_APTX_ADAPTIVE_Q2Q                     (0x00)
+#define A2DP_APTX_ADAPTIVE_NQ2Q                    (0x01)
+#define A2DP_APTX_ADAPTIVE_RESERVED0               (0x00)
+*/
+#define A2DP_APTX_ADAPTIVE_CHANNELS_MONO          (0x01)
+#define A2DP_APTX_ADAPTIVE_CHANNELS_STEREO        (0x02)
+#define A2DP_APTX_ADAPTIVE_CHANNELS_TWS_STEREO    (0x04)
+#define A2DP_APTX_ADAPTIVE_CHANNELS_JOINT_STEREO  (0x08)
+#define A2DP_APTX_ADAPTIVE_CHANNELS_TWS_MONO      (0x10)
+
+#define A2DP_APTX_ADAPTIVE_TTP_LL_0  0x50
+#define A2DP_APTX_ADAPTIVE_TTP_LL_1  0x64
+#define A2DP_APTX_ADAPTIVE_TTP_HQ_0  0x64
+#define A2DP_APTX_ADAPTIVE_TTP_HQ_1  0x64
+#define A2DP_APTX_ADAPTIVE_TTP_TWS_0 0xFF
+#define A2DP_APTX_ADAPTIVE_TTP_TWS_1 0xFF
+#define A2DP_APTX_ADAPTIVE_EOC0 0x00
+#define A2DP_APTX_ADAPTIVE_EOC1 0x00
+#define A2DP_APTX_ADAPTIVE_EOC2 0xAA
+
+#define A2DP_APTX_ADAPTIVE_OTHER_FEATURES_NONE     (0x00000000)
+#define QCOM_AV_APTX_ADAPTIVE_AUDIO                (0x00)
+#define QCOM_CODEC_APTX_ADAPTIVE_ID                (0xff)
+#define A2DP_APTX_ADAPTIVE_CHANNEL                 (0x0001)
+#define A2DP_APTX_ADAPTIVE_SAMPLERATE              (0x22)
+#define CHANNEL_MODE_BACK_CHANNEL_MASK             (0xE0)
+
+#endif  // A2DP_VENDOR_APTX_ADAPTIVE_CONSTANTS_H
diff --git a/stack/include/a2dp_vendor_aptx_adaptive_encoder.h b/stack/include/a2dp_vendor_aptx_adaptive_encoder.h
new file mode 100644
index 0000000..ac127de
--- /dev/null
+++ b/stack/include/a2dp_vendor_aptx_adaptive_encoder.h
@@ -0,0 +1,80 @@
+/******************************************************************************
+    Copyright (c) 2018, The Linux Foundation. All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted (subject to the limitations in the
+    disclaimer below) provided that the following conditions are met:
+
+       * Redistributions of source code must retain the above copyright
+         notice, this list of conditions and the following disclaimer.
+
+       * Redistributions in binary form must reproduce the above
+         copyright notice, this list of conditions and the following
+         disclaimer in the documentation and/or other materials provided
+         with the distribution.
+
+       * Neither the name of The Linux Foundation nor the names of its
+         contributors may be used to endorse or promote products derived
+         from this software without specific prior written permission.
+
+    NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE
+    GRANTED BY THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT
+    HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
+    WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+    MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+    IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+    ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+    GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+    IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+    OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+    IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+//
+// Interface to the A2DP aptX-adaptive Encoder
+//
+
+#ifndef A2DP_VENDOR_APTX_ADAPTIVE_ENCODER_H
+#define A2DP_VENDOR_APTX_ADAPTIVE_ENCODER_H
+
+#include "a2dp_codec_api.h"
+#include "osi/include/time.h"
+
+// Loads the A2DP aptX-adaptive encoder.
+// Return true on success, otherwise false.
+bool A2DP_VendorLoadEncoderAptxAdaptive(void);
+
+// Unloads the A2DP aptX-adaptive encoder.
+void A2DP_VendorUnloadEncoderAptxAdaptive(void);
+
+// Initialize the A2DP aptX-adaptive encoder.
+// |p_peer_params| contains the A2DP peer information.
+// The current A2DP codec config is in |a2dp_codec_config|.
+// |read_callback| is the callback for reading the input audio data.
+// |enqueue_callback| is the callback for enqueueing the encoded audio data.
+void a2dp_vendor_aptx_adaptive_encoder_init(
+    const tA2DP_ENCODER_INIT_PEER_PARAMS* p_peer_params,
+    A2dpCodecConfig* a2dp_codec_config,
+    a2dp_source_read_callback_t read_callback,
+    a2dp_source_enqueue_callback_t enqueue_callback);
+
+// Cleanup the A2DP aptX-adaptive encoder.
+void a2dp_vendor_aptx_adaptive_encoder_cleanup(void);
+
+// Reset the feeding for the A2DP aptX-adaptive encoder.
+void a2dp_vendor_aptx_adaptive_feeding_reset(void);
+
+// Flush the feeding for the A2DP aptX-adaptive encoder.
+void a2dp_vendor_aptx_adaptive_feeding_flush(void);
+
+// Get the A2DP aptX-adaptive encoder interval (in milliseconds).
+period_ms_t a2dp_vendor_aptx_adaptive_get_encoder_interval_ms(void);
+
+// Prepare and send A2DP aptX-adaptive encoded frames.
+// |timestamp_us| is the current timestamp (in microseconds).
+void a2dp_vendor_aptx_adaptive_send_frames(uint64_t timestamp_us);
+
+
+#endif  // A2DP_VENDOR_APTX_ADAPTIVE_ENCODER_H
-- 
2.7.4

